MSC1767: Extensible event types & fallback in Matrix (v2) (#1767)
matrix-org/matrix-spec-proposals: proposals/1767-extensible-events.md

to try and render unknown `msgtype`s.

Note that this fallback only applies in cases where the `m.room.message` `msgtype` was converted to a dedicated
primary event: new features, or events which don't have a `msgtype`, should send their primary event type instead.

### State events

State events are typically not rendered as they usually contain metadata for the room rather than information for
the historical record, however there are a few cases where rendering a state event might be desirable. Clients are
not required to look for event types in state event contents, though are recommended to do so.

This proposal does not introduce any additional requirement for state events to include these rendering details.

A hypothetical scenario would be a bridge state event that includes `m.text` to say a bridge has been added/edited.

### Hypothetical IOT event example

with text fallback:

```json5
{
    "type": "net.arasphere.temperature",
    "content": {
        "m.text": "The temperature is 37C",
        "m.html": "The temperature is <font color='#f00'>37C</font>",
        "net.arasphere.temperature": {

>             "temperature": 37.0

Reviewer: ```suggestion
            "temperature": 37
```

I'm fairly sure that floats are still banned

Reviewer 2: it's also a hypothetical

# MSC1767: Extensible events in Matrix

##  Problem

 1. There is no formal mechanism of extending events with additional structured metadata.
 2. New events tend to reinvent the wheel rather than being able to reuse existing types.

>  3. Clients don’t know how to render unknown event types.

Reviewer 2: Where it does solve case 1 is by standardizing what certain data structures would look like, such as location data, which can be appended to any event to show a "this was sent [here]" sort of UX. Largely though, yes, anyone could shove a `com.example.coords` field onto an event and be fine, but that might not render correctly on all clients in that deployment.

>  3. Clients don’t know how to render unknown event types.

Reviewer 2: I've ultimately rewritten the description to better describe this. Please open a new thread if this feels unresolved :)

Events still have a primary type, which is used when sending them in Matrix.

The names of the keys in contents are now expected to be namespaced, given they
typically refer to types of events.  However, for compatibility, we can keep using
some of the original fields too.

Certain types (e.g. `m.message`, `m.caption`, `m.thumbnail`) which can contain
multiple alternative payloads in different formats express the alternatives as
an ordered list; most preferred first.

We provide short-form types (`m.text` and `m.html`) for the most common scenarios
of an event requiring a plain-text or HTML-formatted representation.

For convenience and compatibility, we keep "body" and "formatted_body" as
shorthand for plain-text and html-formatted fallback for events.

The various types in an event's contents MUST refer to the same event.  Multiple
events should be linked using a `m.relates_to` reference rather than multiplexing
into a single event.

Here are proposals for different types of events:

### m.text and m.html

For compactness for this super-common case, we define a short form which is

> equivalent to the longer form `m.message` example below:

Author: > "text and HTML are special and should be allowed on ANY event type and hence don't follow these mixin rules

this is precisely the intention.

> equivalent to the longer form `m.message` example below:

Reviewer 2: I believe I've clarified this in a refactoring of some of the MSC. If not, please open a new thread.

```json
{

>     "type": "m.message",

Reviewer 3: Is that conflating content-type or MIME-type with, shall we say, event-type or event-purpose?  Or does it imply a default or base event type of `m.message`/simple-text-message?

>     "type": "m.message",

Author: @non-Jedi i covered this in one of the other threads, but from my perspective, it'd valid to have an m.message event with a null content (other than the fallback reps). it'd also be technically valid to have an m.html event (treating the fallback as the primary type), but a bit confusing imo - it'd be nicer if clients just looked for `m.messages` for IMs rather than having to know that some clients disguise IMs as `m.text` or `m.html` for reasons unknown.  So whilst it's valid, i'd suggest recommending against it.

@alphapapa yes, I guess it's kinda conflating content-type with event-purpose. so let's not do it.

>     "type": "m.message",

Reviewer 2: I'm not entirely convinced that we should let the primary type support an `m.html` and `m.text` - `m.message` with appropriate `content` is just as complex to render by a client.

    "content": {
        "m.text": "i am a *fish*",

>         "m.html": "i am a <b>fish</b>"

Reviewer 4: Could we write
```
        "m.text": "i am a fish",
        "m.markdown": "i am a *fish*",
        "m.html": "i am a <b>fish</b>"
```
instead?

Reviewer: I would recommend making a distinction between CommonMark and GFM Markdown (or any other variant), to align converstion

Reviewer 2: Markdown isn't really safe to include at the spec level whereas HTML can be defined very nicely. Maybe once we solve markdown in the spec more generally, but for now I'm extremely hesitant to open the can of worms that is markdown.

>         "m.html": "i am a <b>fish</b>"

Reviewer 4: Also the problem with that approach is that it duplicates the content of the message itself. For a small sentence like `i am a fish` it is ok, but for long text, it could explode the size of the Event.
Maybe we could separate the text itself and the formatting for instance:
```json
{
   "m.text": "i am a fish",
   "m.tags": [
      {
         "start":7,
         "end":10,
         "type":"b"
      }
   ]
}
```
`b` is for bold, and any other tags could be supported, with extra key (for instance `link` with extra key `href`, etc.)
It can then be rendered by the clients at runtime, using the preferred format (Html for Web clients, SpannableString for Android, etc.), and ignoring what is not supported.

Reviewer 5: In addition, the duplication gets worse with internationalisation.

Might as well merge them together into a rich text type, like so.

```json
{
    "type": "m.message",
    "content": {
        "m.rich_text": [
            {
                "type": "plain",
                "body": "i am a "
            },
            {
                "type": "bold",
                "body": "fish"
            }
        ]
    }
}
```

Simple clients can just concatenate all the `body`s together.

Although, nested tags complicates things. e.g. `<b><a href=".....">fish</a></b>`

Reviewer 3: IIUC, both of these ideas would mean that the event would not contain a simple string representation of the whole message.  Clients would have to assemble the pieces just to output a whole, plain-text message.

I can't exaggerate how painful this would be for client authors.  The Internet was built on simple, plain-text protocols, like HTTP, HTML, IRC, etc.  Please do not discard the wisdom inherited from those traditions.  

If you're concerned about size, plain text, Markdown, and HTML all compress well.

Speaking for myself, as the author and maintainer of a somewhat niche Matrix client: if Matrix events were like this, I would never have bothered.  The protocol is already complicated to deal with in many ways.  At some point, dealing with it stops being fun and becomes tedious work, work for which authors of free clients are not compensated.  Please, keep it simple.

Reviewer 4: There is a simple text version in my idea.
Also good compression ratio will not concern encrypted events, so I maintain the idea that we should not duplicate the text itself. Another argument: the text could be different for each format!

```json
"content": {
        "m.text": "i am a *fish*",
        "m.html": "i am a <b>dog</b>"
}
```
Different clients would not display the same text in this case.

This is already the case today and I'm not comfortable with that.

Reviewer 6: Encrypted events can compressed before encryption (with some care).

Reviewer 2: Compressible formats sounds like something that would be best covered by another MSC. The conversation this invokes ends up being long and complicated, and thus ideal for a dedicated MSC.

Reviewer 2: (marking as resolved to indicate the concern has been addressed)

    }
}
```

> ### m.message

Reviewer 2: um, emotes and notices have gone missing.

m.message describes a simple textual instant message.

```json
{
    "type": "m.message",
    "content": {

>         "m.message": [

Reviewer 7: Better still, does `m.message` need to exist at all?

>         "m.message": [

Reviewer 8: Does `m.text` and `m.html` imply that the event is a instant message? I thought that was the point of `m.room.message` (and now `m.message`?)

>         "m.message": [

Reviewer 7: both m.text and m.html imply human readability. What additional properties are meant to be communicated by m. message?

>         "m.message": [

Reviewer 8: Not much that I can think of, honestly.

>         "m.message": [

Author: so the reason to have 
```json
"type": "m.message",
"contents": {
    "m.html": "i am a <b>fish</b>",
    "m.text": "i am a *fish*"
}
```
rather than 
```json
"type": "m.message",
"contents": {
    "m.message": [
        {"m.html": "i am a <b>fish</b>"},
        {"m.text": "i am a *fish*"}
    ]
}
```
is purely to save bits and complexity for the common case. I don't see a problem with considering `m.html` and `m.text` as fully fledged first-class types, which just happen to take a string as their body.  Whether you label the event itself as an `m.message` (which happens to have empty contents for the `m.message`, but instead provides `m.text` and `m.html` fallback event representation) or an `m.html` would be up to you, although I'd suggest `m.message` for sanity.  If you called it `m.text` then you'd expect the client to prioritise the `m.text` payload over the `m.html` fallback.

Conversely, the reason to have
```json
"type": "m.message",
"contents": {
    "m.message": [
        {"m.html": "i am a <b>fish</b>"},
        {"m.text": "i am a *fish*"}
    ]
}
```
over
```json
"type": "m.message",
"contents": {
    "m.message": [
        {"mimetype": "text/html", "body": "i am a <b>fish</b>"},
        {"mimetype": "text/plain", "body": "i am a *fish*"}
    ]
}
```
is to support arbitrary mimetypes without having to somehow map them all onto `m.*` style namespaces.  This is why `m.message` is needed for the general case.

I agree this is a little arbitrary and prone to bikeshedding, but does this make more sense.

To recap:
 * The idea of `m.text` and `m.html` is to compactly describe the types of the common text fallback representations that all events can use.  They are to all intents and purposes normal types, albeit used strategically to provide fallbacks rather than used as a primary event type.  (Similarly you could have something like `m.nsfw` as an NSFW tag which would be a normal type, but very unlikely to be ever used as a primary event type).
 * The idea of `m.message` is to describe an instant message. The full contents of that message could be described by the fallback representations - or one could flesh it out in detail with an `m.message` type (and include additional stuff like the original markdown or latex representations etc).

Does that resonate any better?

>         "m.message": [

Reviewer 8: I wonder if we really want to limit it to just `m.text` and `m.html` for the IOT use case. That's probably one of the best examples of wanting to save bits, and if you are just wanting to send simple values into a room, you could easily end up with:

```JSON
"type": "uk.half-shot.temperature",
"contents": {
    "uk.half-shot.temperature.celsius": 46.5,
}
```

which makes me wonder about @non-Jedi 's proposal of representing these shorthand forms in state, as that would undoubtedly be useful for this use-case.

>         "m.message": [

Reviewer 3: Seems to me that, for custom event types, interpretation should be up to the client.  So, in that example, why not just use:
```json
"type": "uk.half-shot.temperature",
"contents": 46.5
```
Or if it needed to send more than one value, why couldn't it just do:
```json
"type": "uk.half-shot.temperature",
"contents": {
    "temperature.celsius": 46.5,
    "something-else": 0.72
}
```
IOW, why would each key need to be fully namespaced for custom events?

>         "m.message": [

Author: @non-Jedi, no i don't think i flipped them around.

@Half-Shot: long event keys are not too bad as they compress excellently. But adding in structure that otherwise wasn't there at all probably isn't great, hence trying to avoid it.

@alphapapa: yes, one could do completely custom event formats like that, but then you lose the whole ability to supply fallbacks, which is what we were trying to provide, and so we end up back at square 1...

>         "m.message": [

Reviewer 3: @ara4n Maybe I'm misunderstanding the intended use-case then.  In this example of what appears to be an IOT thermometer which uses a custom event type, is a fallback type necessary?  I wouldn't expect (or necessarily want) standard clients, like Riot, to display a message for such events, at least, not by default.  If I needed to see such events, e.g. while debugging the IOT device, I would probably want Riot to simply display raw events in the event list.

From the IOT device's programmer's perspective, I imagine he would want requirements to be as minimal as possible, i.e. the most minimal event, without having to provide fallbacks for clients that aren't the intended target of the event.

I guess I'm missing something.  Sorry if there's context I haven't seen.

>         "m.message": [

Reviewer 8: @alphapapa It might be specific to your idea, but certainly I'd *like* to see fallback text for any event that my client doesn't natively understand for debugging/fun purposes. Of course, you might want to configure your device to not send fallback in really constrained environments (which is also okay, you would do something similar to what we already have).

The major advantage of this proposal is you can write thermometer events that have optional text content to display in a client in a prettier format than raw data.

The *other* advantage is you could technically have two types of non-renderable content: One could be a generic sensor type which would show up on say, a dashboard app. And then you could have a branded event type which would contain extra info about your specific sensor that might only apply to a specific app.

>         "m.message": [

Reviewer 8: Also it would be nice to finally be able to do things like have Twitter tweets under their own event type while still being render-able in $client. At the moment you have to make things into a `m.room.message` or they won't show up at all :/.

>         "m.message": [

Reviewer 2: (low bandwidth, such as IOT, is probably best handled by dedicated MSCs like https://github.com/matrix-org/matrix-doc/pull/3079)

>         "m.message": [

Reviewer 2: I *think* this thread is resolved, but if not then a more concise concern in a new thread would be appreciated.

            {
                "mimetype": "text/html",
                "body": "i am a <b>fish</b>"
            },
            {
                "mimetype": "text/plain",
                "body": "i am a *fish*"
            }
        ],
    }
}
```

> ### m.file

Reviewer 9: Why do we need other types than `m.message`? Why can't an m.message just contain a file / image / whatever?

Author: The proposal is that the event type describes the type of the data in that event.  So an instant message would be an `m.message`; an image would be an `m.image`; a file would be an `m.file`, so the typing is more useful and descriptive and you can filter based on it.  It means you use the same mechanism for displaying a fallback for a custom event type as a normal event type.

Reviewer 10: In most cases, separate images/files are not useful without context (text description), [here](https://github.com/matrix-org/matrix-doc/pull/2529#issuecomment-736373314) is more detailed description of this problem.

And when we found the event with interesting image, we must immediately find the text event, that referenced to this image (describes it). So why not to store this image directly in that event, instead of have 2 references (`m.message` > `m.image` > mxc url to image).

At wider look, current `m.image` and `m.file` events are already not actual events, but references to file via `mxc` url, so why we must use double references to simply insert image inside text? ;-)

Reviewer 2: I believe Matthew's comment addresses this thread. Please open a new thread if feeling otherwise.

```json
{
    "type": "m.file",
    "content": {
        "m.text": "foo.dat (12KB)",
        "m.file": {
            "url": "mxc://matrix.org/asdjkhcsd",
            "name": "foo.dat",
            "mimetype": "application/octet-stream",
            "size": 12345
        }
    }
}
```

### m.image

```json
{
    "type": "m.image",

>     "content": {

Reviewer 10: I think that use array instead of object in `content` will give much more flexibility to current implementation and future improvements, because will allow to add several items of same type. But simple change of object to array will broke fallback, so we can select new key for it, something like `contents` or `items`, `elements`.

And for fallback - we will keep `content` object and put in it the rendered aggregated fallback display of all complex event content items, here is example:
```json
{
  "type": "m.message",
  "contents": [
    {
      "msgtype": "m.text",
      "body": "Invite to Party at Bob's House",
      "format": "org.matrix.custom.html",
      "formatted_body": "Invite to Party at <i>Bob's house</i>"
    },
    {
      "msgtype": "m.calendar.request",
      "version": "2.0",
      "summary": "Party at Bob's house",
      "vEvent": {
        "dtStart": {
          "date": "20180207T150000",
          "tzId": "Europe/London"
        },
        "dtEnd": {
          "date": "20180207T180000",
          "tzId": "Europe/London"
        }
      }
    },
    {
      "msgtype": "m.file",
      "uri": "mxc://matrix.org/auiheiuh1eqwd",
      "mimetype": "text/calendar",
      "name": "bobs-party.ics",
      "size": 1246
    },
    {
      "msgtype": "m.location",
      "uri": "geo:37.786971,-122.399677;u=35",
      "description": "123 Fake Street - first part"
    },
    {
      "msgtype": "m.location",
      "uri": "geo:38.786971,-123.399677;u=35",
      "description": "234 Real Street - second part"
    },
    {
      "msgtype": "m.image",
      "uri": "mxc://localhost/JWEIFJgwEIhweiWJE",
      "info": {      
        "width": 256,
        "height": 256,
        "mimetype": "image/jpeg",
        "size": 1024000
      }
    }
  ],
  "content": {
    "msgtype": "m.text",
    "body": "Invite to Party at Bob's house\nDate: Feb 7th, 3pm\nLocations: [37.786971,-122.399677;u=35] (123 Fake Street - first part); [geo:38.786971,-123.399677;u=35] (234 Real Street - second part)\nFile: mxc://matrix.org/auiheiuh1eqwd>bobs-party.ics\nImage: mxc://localhost/JWEIFJgwEIhweiWJE",
    "format": "org.matrix.custom.html",
    "formatted_body": "Invite to Party at <i>Bob's house</i>\nDate: Feb 7th, 3pm\nLocations: [37.786971,-122.399677;u=35] (123 Fake Street - first part); [geo:38.786971,-123.399677;u=35] (234 Real Street - second part)\nFile: <a href=mxc://matrix.org/auiheiuh1eqwd>bobs-party.ics</a>\nImage: <img src=mxc://localhost/JWEIFJgwEIhweiWJE />"
  }        
}
```
What do you think about this idea?

Reviewer 10: For help Matrix clients to understand which version to display (try to render each item by himself, or simply display the fallback text), we can also add the item type, that describe which features client must support for properly render all by himself, something like `features_used: ['MSC1234', 'MSC2345']` or `client_server_version_used: r0.7.2`.

Reviewer 10: This way also can be confused with allowing several items of same type (for example, 3 `m.text` items), but we can handle this via joining text of all `m.text` items into one `contents.body` field, and have similar solutions for other types.

Author: My concern about letting people bundle together load of different items in a single event is:
 * It encourages folks to bundle unrelated things together, making events non-atomic
 * You have no way of linking to the component parts
 * It would complicates the spec
 * It makes the events big and encourages people to hit up against the 65K limit
 * Relations *should* provide a much more flexible way of bundling together a set of related events

Reviewer 10: > It encourages folks to bundle unrelated things together, making events non-atomic

Any way all clients must render each event as full single "atom" of timeline, so aggregating it from many events will be superfluous routine.

I can't present the situation, when we need to show only part of event (excluding media item content, that is already the relation to mxc). 

All items of the event is the parts of it, and must not be considered as independent events. 

> You have no way of linking to the component parts

Yes, but situations when we need link to some part of event are very rare. And, if very needed, we can link to numeric serial number of needed item in array.

Sometimes we need to quote some paragraph of event text to reply on event, but this isn't means that we must split the text paragraphs to separate items. 

> It makes the events big and encourages people to hit up against the 65K limit

Yes, but pasting long text in simple m.message event can break this limit too ;)

> Relations should provide a much more flexible way of bundling together a set of related events

Relations are not disappearing, each relation can be part of current event array, but in needed place (eg between two text items), instead of the separate array like now (that make impossible to insert one needed relation in middle of event, and others - at end of it).

Reviewer 11: > It would complicates the spec

Slack, Discord, Guilded and Facebook have similar rich embeds and they handle it just fine.
> It makes the events big and encourages people to hit up against the 65K limit

Jumbo event support may be added if required.

Reviewer 2: At a glance this thread looks like discussion around display hints and multiplexing, which has zero interest in being included in this MSC. Other MSCs are welcome to argue for their inclusion.

        "m.text": "matrix logo (640x480, 12KB)",
        "m.image": {
            "width": 640,
            "height": 480,
        },
        "m.file": {

>             "url": "mxc://matrix.org/asdjkhcsd",

Author: The reason I didn't is because the event should be processable as a plain old `m.file` by a client that doesn't know how to handle `m.image` (or `m.video` or `net.arasphere.weirdformat`).  But you're right that the field duplication is bugging me.

Perhaps the right solution is to treat `m.thumbnail` as a nested event, and embed within it its own `m.file` and `m.image`.  The only reason i didn't was to special-case for simplicity, but with a fresh pair of eyes it certainly looks a bit jarring.

>             "url": "mxc://matrix.org/asdjkhcsd",

Author: actually, if we treat `m.thumbnail` as nested events then we get stuck into whether they need captions or not, etc.  having them as a special-case (rather than heading off re-creating displayhints) feels like the least evil.

>             "url": "mxc://matrix.org/asdjkhcsd",

Reviewer 2: I've tried to make the fields a bit less duplicated in https://github.com/matrix-org/matrix-doc/pull/3552

            "mimetype": "image/jpeg",
            "name": "logo.jpg",
            "size": 12345
        },

>         "m.caption": [

Author: because arrays are for lists of fallback format representations of the same type (text caption, html caption, etc)

>         "m.caption": [

Reviewer 2: ah, right. Okay. Might be worth having an example of that (unless I'm blind and there is one)

            "temperature": 37.0
        }
    }
}
```

### m.calendar.request

This is a deliberately chunky event designed to show a geotagged calendar request.

(In practice we should probably use the IETF jsCalendar format rather than inventing a
new one here, and to ensure interop with the JMAP folks.
https://tools.ietf.org/html/draft-ietf-calext-jscalendar-11)

```json
{
    "type": "m.calendar.request",
    "content": {
        "m.text": "Invite to Party at Bob's House on Feb 7th, 3pm",
        "m.html": "Invite to Party at <i>Bob's house</i> on Feb 7th, 3pm",
        "m.calendar.request": {
            "version": "2.0",
            "summary": "Party at Bob's house",
            "vEvent": {
                "dtStart": {
                    "date": "20180207T150000",
                    "tzId": "Europe/London"
                },
                "dtEnd": {
                    "date": "20180207T180000",
                    "tzId": "Europe/London"
                }
            }
        },
        "m.file": {
            "uri": "mxc://matrix.org/auiheiuh1eqwd",
            "mimetype": "text/calendar",
            "name": "bobs-party.ics",
            "size": 1246
        },

>         "m.location": {

Reviewer 7: So clients are expected to compute (based on what?) which keys are for fallback representation and which are additional information?

I don't think heuristics get you far enough in figuring this out. Maybe instead have a `fallbacks` field at the same level as `content` containing a struct of keys=fallback-type, value=fallback-value; the expectation would be that the client would choose exactly one type from the `fallbacks` struct rather than being able to display multiple related types like under the `content` key.

Author: the idea is that yup, it'd be heuristics for now and might even be non-deterministic between clients. if you want more control, it'd be via display hints, which i'm going to reintroduce in another MSC to try to split them out from the core question of extensibility.

Reviewer 7: I missed Travis's comments to this same effect below. Feel free to mark this one as resolved.

            "uri": "geo:37.786971,-122.399677;u=35",
            "description": "123 Fake Street"
        },
        "m.thumbnail": [
            {
               "uri": "mxc://localhost/JWEIFJgwEIhweiWJE",
               "width": 256,
               "height": 256,
               "mimetype": "image/jpeg",
               "size": 1024000
            }
        ],
    }
}
```

This would be nicer with displayhints support, but still works pretty well.

## Tradeoffs

It's a bit ugly to not know whether a given key will take a string, hash or array.

> It's a bit arbitrary as to which fields are allowed lists of fallbacks.

Reviewer 7: To generalize this just a bit, instead of defining a `type` field which is primary display type at same level as `content`, we could have `content` itself contain a fallback list of the format:

```json
{
  "content": [
    {"m.html": "i am a <b>fish</b>"},
    {"m.text": "i am a *fish*"}
  ]
}
```

That way within the `content` structure lists would always indicate ordered fallbacks, and we could define fallback priority rather than just what the best type to display is.

Author: the reason I avoided doing it this way is that you can't list fallbacks of the same type - e.g. two different `m.thumbnails` (one larger and animated; the other smaller and boring).

Reviewer 7: Why couldn't you?

```json
{
  "content": [
    {"m.html": "i am a <b>fish</b>"},
    {"m.html": "Why would you <b>want</b> to have two separate html messages? Who knows?"},
    {"m.text": "i am a *fish*"}
  ]
}
```

This still might not be a good proposal for other reasons, but I don't see this particular criticism.

Author: oh, yeah, sorry - i thinkoed. so yes, i think the difference between:

```json
{
  "content": [
    {"m.html": "i am a <b>fish</b>"},
    {"m.html": "Why would you <b>want</b> to have two separate html messages? Who knows?"},
    {"m.text": "i am a *fish*"}
  ]
}
```

and

```json
{
  "content": {
    "m.html": [
        "i am a <b>fish</b>",
        "Why would you <b>want</b> to have two separate html messages? Who knows?"
    ],
    "m.text": "i am a *fish*"
  }
}
```

is fairly subjective. I personally prefer putting the lists per-key rather than muddling them all together in a backwards-incompatible way, but i guess it means you need prior knowledge (or implement displayhints) at how to order types with respect to one another.

Hum.

Reviewer 2: Changing `content` to be an array sounds extremely complicated from a spec perspective - we favour objects for their extensibility, and the content is already an event. Making it suddenly an array would break all sorts of details within servers and clients.

The `content` array does look cleaner, but I don't think that's enough of a justification to break the protocol at this point in time. A future MSC is welcome to figure this out.

It's a bit ugly that you have to look over the keys of contents to see what types
are present, but better than duplicating this into an explicit `types` list (on balance).

> We're skipping over defining rules for which fallback combinations to display

Author: unconvinced, especially with the ordered lists for m.message, m.caption and m.thumbnail. grouping the contents fields into disparate types is a massive win for reuse and extensibility; having displayhints for determining when to display weird and wonderful combinations of types is the icing on the cake (and really *does* make this more complex)

> We're skipping over defining rules for which fallback combinations to display

Reviewer 2: I'm not convinced that the features introduced by ordered lists are solving a real problem. I also don't see the benefit of extensible events if I can't create my own events and have them rendered. As far as I can tell, this proposal doesn't include any kind of rendering rules for when a client runs into a non-standard event type.

> We're skipping over defining rules for which fallback combinations to display

Author: So if the client doesn't recognise the event type, it'd use m.message/m.text/m.html, m.thumbnail and m.file to synthesise a plausible fallback.  You're right that this needs to be explicit; something like: "if the primary type isn't recognised, show the m.message, and m.file link if present and an m.thumbnail if present".  I'd be interested if we have common use cases which wouldn't be sufficiently covered by this.

(btw i definitely think displayhints should be proposed as another MSC; just don't want this one getting stuck on it).

> We're skipping over defining rules for which fallback combinations to display

Reviewer 2: The proposal should definitely spell that out as it's not entirely clear what the client should be doing when it receives such an event. The common case is probably resolved with this kind of fallback, but I can think of a couple use cases which may benefit from richer parsing rules (journal and matrix-static immediately come to mind).

> We're skipping over defining rules for which fallback combinations to display

Author: can you give some sample events for the journal & matrix-static use cases so i can see what you have in mind? my hunch is still that we should make display hints a separate thing

> We're skipping over defining rules for which fallback combinations to display

Reviewer 2: They'd be more on the rendering side of things and could potentially benefit from being told what to do. 

It's probably fine to punt display hints to its own MSC - it looked like this proposal didn't have a rendering mechanism and I really liked how display hints worked in v1.

(i.e. "display hints") for now; these can be added in a future MSC if needed.
MSC1225 contains a proposal for this.

## Issues

I think Erik had some concerns about mixing together types at the top level of `contents`
but I've forgotten the details.  Hopefully these are mitigated by the revised approach.

## Security considerations

> We can't apply ACLs serverside to the types embedded in the event contents.

Reviewer 3: This is a tough call IMO, because it potentially adds complexity to clients.  They could render events unconditionally, but, to a certain extent, that would make them vulnerable to trolls.  Imagine a room in which only users with a certain PL could post images, and clients were responsible for enforcement: if a certain client had a bug in its conditional rendering, and all the room's mods were running clients that worked correctly, trolls could post messages that would look innocuous to mods, but users with buggy clients would see whatever the troll put in the event's other content types.  Probably not a serious concern, but something to consider.

While I am reluctant to add more complexity to clients, it probably makes sense to be consistent across all room types, to the extent practical.

> We can't apply ACLs serverside to the types embedded in the event contents.

Author: So, i totally agree that a typical client developer shouldn't have to worry about E2EE.

However, I am increasingly wondering if in the longer term (like 2-3 years away) we may find ourselves E2EEing everything everywhere.  The reasons are:
 * Dogfooding E2E;
 * Mitigating legal risk for server admins if some obnoxious content from a public room ends up on their server.
 * We start to see publicness as more of a spectrum - just because a room is theoretically open to anyone from the public, it doesn't mean that everyone knows how to find it, or that its contents should be unencrypted.
 * I really hope that we will provide a matrix daemon service which can offload all the E2E lifting from a typical Matrix client.  So a typical implementor will talk CS API to the daemon listening on localhost, which dutifully sits there encrypting/decrypting and proxying through to your actual server. In future, the server may also optionally sit clientside. At which point E2E becomes much more prevalent.  Such a daemon could also offload the finicky business of enforcing event ACLs clientsid.

In practice, though, I honestly don't think that enforcing ACLs clientside is /that/ big a deal.  It's the sort of thing that you skip in the first implementation and then find yourself adding the first time someone abuses it.

As a result, I feel that designing features which only work on non-E2E feels is a bit of a bad smell.

Would be interested in what others of @matrix-org/spec-core-team think tho.

> We can't apply ACLs serverside to the types embedded in the event contents.

Reviewer 2: for the record, 2-3 years later we already have use cases where entire servers are encrypting every room they create, sometimes to the detriment of usability for the sake of security (tens of thousands of devices in multitudes of rooms). Clients already have to do client-side validation for encrypted events, and typically make decisions on what to render, only to change that when something in the environment changes (as Matthew mentions). 

We also have a daemon now, making a lot of the work offloaded for typical clients and even most bots.

It's a couple years late, but my feeling as a SCT member is that we should definitely not exclude encryption. It's critical to the protocol now.

> We can't apply ACLs serverside to the types embedded in the event contents.

Reviewer 6: I agree. E2EE should be the first priority of any design without an explicit reason why it shouldn't. This is because non-encrypted rooms are almost always automatically handled by any design that works for E2EE and we don't want to have two specs for each feature (that would put a real burden on clients).

> We can't apply ACLs serverside to the types embedded in the event contents.

Reviewer 2: I plan to move a lot of this discussion to a future MSC with placeholder https://github.com/matrix-org/matrix-doc/issues/3556 - resolving until then.

However, this is inevitable given the existence of E2E, so we have no choice but
for clients to apply ACLs clientside (e.g. refuse to render an m.image contents
on an event if the sender doesn't have enough PL to send an m.image event).

## Availability

> Given the disruption of migrating clients & bridges to the new event shape, this

Reviewer 8: The migration path definitely needs to be known before this can land in spec, because I'd like for at least all the major SDKs and clients to be quick on the uptake to avoid everyone clinging to `m.room.message` as the most compatible option.
My suggestion would be for SDKs to start sending `m.message` in helper functions, and *by default* they should translate `m.message` messages into `m.room.message` where possible. Almost all the JS based bridges rely on this consistent format and it would be good to let developers gradually make changes to their code rather than suddenly expecting them to adapt.
Hopefully after 2-3 months a new breaking version could be written for the JS-SDK which makes `m.message` the default everywhere which would convert inbound `m.room.message`s into `m.message`.

> should probably land after S2S r0.

Reviewer 8: Sidenote: Does changing the format bugger up push rules at all?

Reviewer 2: it makes the default ones a bit annoying - we should open a separate MSC to handle that.

Reviewer 2: punted to https://github.com/matrix-org/matrix-doc/issues/3555 as a future MSC.

> # MSC1767: Extensible events in Matrix

Reviewer 2: @alphapapa says:

> Revisiting this after reading [Andy's blog post](https://www.artificialworlds.net/blog/2022/03/08/comparison-of-matrix-events-before-and-after-extensible-events/), I have some concerns again.
> 
> As a minor client developer, this feels discouraging.  Now, instead of writing code to parse messages in one structure, I have to parse messages in:
> 
> 1.  The original specified structure.
> 2.  The new, extensible structure with multi-keyed maps in a list of maps.
> 3.  The new, extensible structure with single-keyed abbreviations in a map that is not in a list of maps.
> 4.  A potential mishmash of 3 different structure specifications shoved into a single event.
> 
> So instead of straightforward destructuring code, I have to write a long conditional trying four different ways.  And all of those will have to be kept around in perpetuity, or else someday, old events might just appear empty or, worse, cause bugs that might make my client non-functional.
> 
> As well, the wire format gets incredibly bloated.  Look at the last example given: the same message body written four different ways in a single event.  This doesn't only affect bandwidth and disk usage, but it also makes debugging harder, because it makes raw events much more difficult to visually parse.
> 
> Matrix is generally great, but stuff like this isn't.  Please rein this in before it goes into spec.  Simplify this down to the bare minimum.  You can always add another MSC later if absolutely necessary.
> 
> Matrix has so much potential, and is based upon a fundamental simplicity (lists and key-value maps), but I fear that the Matrix protocol is going to eventually become so bloated and confusing that developing clients (not to mention servers) is going to become impractical.  Interoperability between clients may suffer as well, and if that happens, it will destroy Matrix's reputation among users and probably its future.  It has plenty of competitors that will be glad to make their specs simpler and easier for developers.  If Matrix wants to be the next 30-year protocol (like HTTP, et al), it needs to keep simplicity as a paramount objective (i.e. think HTTP 1.x, not 2.x/3.x).

Reviewer 2: For clarity: You don't need to support this parsing structure. For clients wishing to try and render *unknown event types* though, this is how they'd do it. For example, when a new feature is being developed and gracefully falls back to text: clients which support extensible events can render it as text until they gain knowledge on the primary event type.

As an example, your client can destructure `m.message` events by reaching into the `content` and picking out the fields it needs, much like today's `m.room.message` handling. This doesn't require interpreting the extensible events schema, by design.

We are also deliberately prioritizing rendering capabilities over readability/bloat of the event here. We want to support many of the features already described by the MSC and have narrowed it down to this as the smallest possible structure to represent the information required. While bandwidth can be argued as a concern, we envision a future MSC will address the transport issues rather than trying to solve it at the event level. 

Generally speaking though, this is already at a bare minimum for acceptance. We moved all the non-text bits out to their own MSCs and have spent a couple years working out how exactly to represent the structure in the simplest and most understandable form, landing on this. We're not saying it's perfect, but it's certainly better than the early approaches - we plan to supplement gaps with blog posts similar to Andy's as we continue to push this out.

The complexity of writing implementations of Matrix is a very well-known problem, and one that we already feel today. It's off topic here, but the plan is to split the spec into logical parts (exact breakdown to be determined - currently looking like events, voip, encryption, etc) so clients and servers can more clearly see what they actually *need* to implement to fulfill a given usecase. A server targeting text-only clients will require encryption but not voip, as an example. We believe that extensible events is net-positive, all things considered.

Reviewer 3: @turt2live I appreciate your thoughtful response.  I do hope this proposal works out well and that Matrix succeeds in the long-term.

Reviewer 12: FWIW, I think the above may be overcomplicating it a bit and there is a simpler way of stating this.

To support the parsing side of this MSC, do this:

- If you see a `m.message` event, just parse the new format. Don't support the old format at all.
- If you see a `m.room.message` event, first try parsing the new format. If that fails, parse the old format.

To support the sending side of this MSC, do this:

- Continue sending `m.room.message` events as normal, but also add in the keys of the new format.
- At some agreed upon point in the future (pending a new MSC), change the type to `m.message` and stop sending keys from the old format.

This doesn't seem so bad to me. If I'm interpreting the MSC correctly, perhaps something like this could be added to a "What you need to do as a client author" section? The MSC talks quite exhaustively about the rationale behind the design and the transition process, but perhaps clear instructions to client authors were a bit lost in the noise.

Reviewer 13: While the actual thread for this is here: https://github.com/matrix-org/matrix-spec-proposals/pull/1767#discussion_r783533194 , there is actually some more complication specifically for text message events, so a client needs to parse these in approximately this order to ge the text content of an event:

- search for mimetype "text/html" in "m.message"
- look at "m.html"
- look at "format" and "formatted_body"
- search for mimetype "text/plain" in "m.message"
- look at "m.text"
- look at "body"

I disagree that this is the simplest way this proposal can be represented, 4 steps should be enough, 6 is too much. But yeah, that's https://github.com/matrix-org/matrix-spec-proposals/pull/1767#discussion_r783533194 .

Reviewer 12: (I took that thread as being more about whether shorthand keys are warranted or not, and that's why I didn't comment there. I don't really have a strong opinion on that topic and I can see benefits to either approach.)

However, I'd argue that's not the correct list of steps due intermixing of individual parsing steps from the new format and the old format. I think this is a distinctly bad idea.

I think the correct way of parsing should always be

1. Parse the new format.
2. If failing 1, parse the old format.

Which when expanded leads to

1. a. Search for mimetype `text/html` in `m.message`.
   b. Look at `m.html`.
   c. Search for mimetype `text/plain` in `m.message`.
   d. Look at `m.text`
2. a. Look at `format` and `formatted_body`.
   b. Look at `body`.

That is, only when we have exhausted all of the new format parsing steps should we even attempt to look at old-style fields. Clients should never send events without `text/html`/`m.html` but *with* `formatted_body`. Parsing it this way ensures that such events will be considered invalid.

Reviewer 3: I don't love everything about Python, but I think the Zen of Python is applicable here: there should be one, obvious way to construct and to parse messages--and preferably only one correct way.  

IMHO, these shorthands don't actually make it easier--they make it harder, because they are additional cases that must be programmed for.  It's much easier to write a program to do one specific thing than to write one to do A, but maybe B, or maybe some of A and some of B, and eventually falling back to C if nothing else works.

So I would suggest picking either the shorthands or the full MIME-types and sticking with it.

While events are currently JSON blobs which accept additional metadata appended to them,
there is no formal structure for how to represent this information or interpret it on the
client side.

New events also typically reinvent the wheel rather than reuse existing types, such as in

> cases where captions, thumbnails, etc need to be considered. This has further issues of

Reviewer 12: What role does this MSC play in captions, given that it's not meant to represent mixins?

Reviewer 2: this MSC? Nothing! :D

https://github.com/matrix-org/matrix-doc/pull/3552 makes a (questionable) argument for captions.

clients not knowing how to render these unknown events, leading to mixed compatibility
within the ecosystem.

The above is seriously hindering uptake of new event types in Matrix - whether that’s
richer data types (stickers, location, calendaring, etc) or IOT-style use cases.
It also means we are currently using an underspecified solution for rich

> messaging, as there has never been a previous agreed way of a jury rig solution
> for rich messaging as we never previously agreed a way of expressing alternate

Reviewer 14: Since I don't understand any of the technical content, some bikeshed-painting.

```suggestion
messaging, as there has never been a previously-agreed way of creating a jury-rigged solution
for rich messaging, as we never previously agreed a way of expressing alternate
```

Reviewer 14: still too many "previously agreed". bah.

* [MSC3551 - Files](https://github.com/matrix-org/matrix-doc/pull/3551)
* [MSC3552 - Images and Stickers](https://github.com/matrix-org/matrix-doc/pull/3552)
* [MSC3553 - Videos](https://github.com/matrix-org/matrix-doc/pull/3553)
* [MSC3554 - Translatable text](https://github.com/matrix-org/matrix-doc/pull/3554)

## Proposal

Events continue having a primary type when sending them in Matrix, but the contents
are grouped together into different types. This allows events to easily extend an
event with additional metadata, re-use existing structures, and have the event
renderable on clients which have not special-cased support for the primary type yet.

The names of the keys in the contents are expected to be namespaced, aligned with
event types, though some keys might not be labeled as such for backwards compatibility.

When types are represented as arrays, they are ordered in preference for rendering.
For example, an `m.thumbnail` might have multiple sizes: a client which does not
support/understand thumbnail sizing would use the first available thumbnail rather
than finding the best possible one.

The event types referenced in the event content MUST refer to the same event and

> cannot be used to multiplex data into a single event. Relationships between events

Reviewer 14: This sentence seems important because of the MUST, but I don't really understand it what is trying to tell me. There are no event ids involved here (afaict), so how does an event type refer to different events?

Reviewer 15: I think that the word "event" in "... MUST refer to the same event ..." does not refer to a Matrix event, but rather some "thing", and perhaps it would be better to change that word to something else.  Some possible words are "item", "thing", "object", "data".

Reviewer 2: This is trying to say "these aren't mixins" in response to the original proposal - it definitely needs rewording having re-read it 😬 

tldr: adding `m.location` to the event content of `m.message` doesn't make it "text with location" as those are two logically different things - a new `m.text_with_location` event type (pending proper name) would be introduced instead.

Reviewer 12: I think just including what you last said would go a long way towards clarity. Something like inserting the following somewhere:

"That is, extensible events are not intended to represent mixins. For example, an event containing both a `m.location` and a `m.message` content does not represent a "text with location". Rather, it's a location with a text fallback for clients which don't understand locations."

There's already an example with an album of images so it might even make sense to make a list of several examples to build intuition.

Also, I think the point about relations which follows might be adding to the confusion since it is introduced a bit too abruptly. I suppose it's trying to say that, if you are actually trying to form a relationship between two events, you shouldn't attempt to use extensible events for this, but rather that you should send a separate event with a `m.relates_to`. But it's not terribly clear that this is indeed what is intended.

Reviewer 2: I think I will have clarified this with an inbound rewrite. If not, please definitely open a new thread.

are formed with `m.relates_to` and similar structures. For clarity, and as an example,
this means that an image event refers to exactly one image - an album of images can
be represented as a thread/relationship chain of singular images.

> Because events are generally sent in the context of a room, all primary types introduced

Reviewer 14: is this a different sort of "primary type" to the "primary type" defined at line 34?

Reviewer 2: yes.

to the Matrix specification are expected to drop their "room" part. For example, instead
of `m.room.event` the type would be `m.event`. Where features have multiple events, such
as polls, the expectation is that they'll declare a relevant namespace like `m.poll.*`.

> ### Text messaging

Reviewer 14: Is this an example of how we're applying the changes described above? or additional changes?

Reviewer 2: This is part of the proposal content to actually make use of the structure outlined above. It felt a bit pointless to say "events are changing!!" without a demonstrated case for it (plus the administrative overhead of the MSC process if we were to break it out).

Currently text content can be identified from the `msgtype` on `m.room.message` events:
`m.text`, `m.emote`, and `m.notice` being the values.

All three additionally support HTML through a combination of `format` and `formatted_body`.

When translated to Extensible Events, the primary type now maps to a slightly different
value:
* A `msgtype` of `m.text` would be a primary type of `m.message`
* A `msgtype` of `m.emote` would be a primary type of `m.emote`
* A `msgtype` of `m.notice` would be a primary type of `m.notice`

All 3 primary types have the same event content structure:

```json5
{
    "type": "m.message",
    "content": {
        // short form

>         "m.text": "i am a *fish*", // doesn't have to be markdown, but useful as a "fallback" for HTML

Reviewer 14: so `m.text` are special cases which are only valid in the content of an `m.message`? maybe not the best example to go in with if so?

Reviewer 2: `m.text` is just a shortcut for `"m.message": [{"mimetype": "text/plain", "body": "..."}]` in `content`

        "m.html": "i am a <b>fish</b>"
    }
}
```
```json5
{
    "type": "m.message",
    "content": {
        // longer form

        "m.message": [
            {
                "mimetype": "text/html", // optional, default text/plain
                "body": "i am a <b>fish</b>"
            },
            {
                "mimetype": "text/plain",
                "body": "i am a *fish*" // doesn't have to be markdown, but useful as a "fallback"
            }
        ],
    }
}
```

> `m.text` and `m.html` are shortform types used to represent the most common scenarios. In cases

Reviewer 15: To be pedantic, I'm understanding the "shortform types" to be interpreted as:

- if both `m.text` and `m.html` are present, then treat it as if there was an `m.message` property with a `{"mimetype":"text/html", "body": ...}` and a `{"mimetype":"text/plain", "body": ...}` entry, in that order
- if only `m.html` is present, then treat it as if there was an `m.message` property with a `{"mimetype":"text/html","body":...}` entry
- if only `m.text` is present, then treat it as if there was an `m.message` property with a `{"mimetype":"text/html","body":...}` entry

the important part being the case where both of the shortforms are present, where we want the HTML version to be preferred over the text version.

Reviewer 2: that's the theory, though in practice `m.html` on its own isn't all that helpful as the client would be depending on `m.text` - there's another thread out there which considers removing `m.html`, and I'm leaning towards doing that for simplicity reasons.

Reviewer 15: > @uhoreg, presumably your last bullet point has a typo and it's meant to say `"mimetype":"text/plain"` instead?

Yes.  Fixed

> `m.text` and `m.html` are shortform types used to represent the most common scenarios. In cases

Reviewer 16: Although it always appears in the examples, it is not mentioned that a plain text representation (`m.text` or `{"mimetype":"text/plain", "body": ...}`) is required when an `m.message` is expected. This diverges from the current spec where the `body` is required on the defined message types. It also means that a client might not be able to rely on the textual representation of a message, even if it is present.

Is this a conscious choice?

I tried to go over the previous comments and I couldn't find a reference to that so I'm sorry if this concern has already been answered.

Reviewer 2: The intention is that `m.message` primary event types have at least a plain text representation of some kind, be it `m.text` or `m.message` in the body.

Reviewer 16: With the current phrasing what I understand is that *any* `m.message` is valid. So not necessarily `text/plain`.

where either `m.text` and `m.html` are specified alongside `m.message`, `m.message` should be

> preferred.

Reviewer 13: If we are not going to map all event types into a top level key like `m.markdown`, `m.json` etc anyway, I can't see a strong reason to have `m.html`. That seems to only add complexity with marginal usefulness. You save less around 10 characters, but you need to include a text representation anyway. I can somewhat understand having a shorthand for `m.text`, because there you can actually save some data for IOT or whatever low bandwidth projects. But what is the argument for having a shorthand for `m.html`? Can we just require that to be in `m.message` with a mimetype of `text/html`?

Reviewer 2: Most clients send out text+html already, so the theory is instead of a really awkward `m.message` array in the middle of their send event code path we can just shorthand it, particularly if the client doesn't mix the generation and parsing concerns.

For clients which do mix the generation and parsing concerns, it's mostly moot to them because they won't see the event content being constructed in their code.

Eg:
```js
client.sendEvent(roomId, "m.message", {
  "m.text": userInput,
  "m.html": htmlUserInput,
});
```

Reviewer 13: That sounds like a nice API, that an SDK would add and not something, that needs to be encoded in the spec. Especially since it makes the parsing quite a bit more ugly, since you always need to parse the array first, otherwise fallback to `m.html`, then fallback to `m.text` and lastly, in the transition phase, fallback to `body`. Especially since statically typed SDKs might want to add specific types for MHtml anyway and handling special cases can be quite awkward.

Reviewer 2: The original thread was https://github.com/matrix-org/matrix-doc/pull/1767#discussion_r245401460 if it helps - open to more feedback/ideas if `m.html` feels spurious (I personally don't really have a preference)

Reviewer 13: Yeah, I read that one, but the primary argument seems to be compactness and that makes sense for `m.text`, but not `m.html` (since an event needs to have a certain minimum length and complexity for it to even include html formatting) and imo we should have as few special cases as possible in the already more complex extensible event. Most SDKs will probably just automatically canonicalize the event somewhat, but still, every special case is just a new opportunity for a new bug.

Reviewer 2: Out of band discussion revealed that `m.html` is a hard blocker for this MSC going forward, but other shortcuts don't really make sense. This will be clarified in an inbound rewrite.

> preferred.

Reviewer 2: correct, yes.

For completeness, here are emotes and notices:

```json5
{
    "type": "m.emote",
    "content": {
        "m.text": "is a fish" // note the semantics of the old msgtype are preserved
    }
}

> ```

Reviewer 15: Is the long-form version of this:

```json5
{
  "type": "m.emote",
  "content": {
    "m.emote": {
      "mimetype": "text/plain",
      "body": "is a fish"
    }
  }
}
```

(where the property in the `"content"` is `m.emote` instead of `m.message`) ?

Reviewer 2: negative, `m.message` in the `content` for the type.

```json5
{
    "type": "m.notice",
    "content": {
        "m.text": "I am a bot who does not wish to use normally-styled text"
    }
}
```

### Rendering (un)known events

A client's approach to rendering a known event is unchanged: if the primary event type is known, the
client will pick out the details it needs from `content` and render accordingly.

If the primary type is not known, the client should inspect `content` for types it does know and render
the best match. Determining the best match is intentionally left as an implementation detail, however
clients might wish to use a modified version of the following sequence:

1. All known primary event types not listed in this list.
  * Some events, like [polls](https://github.com/matrix-org/matrix-doc/pull/3381), might not make sense to
    fall back to. Clients should only use primary types which make sense for their specific application,

>     which might typically be just the ones listed below.

Reviewer 14: this is rendering weirdly, fyi

Reviewer 2: nothing I can really do about this - it's a GFM oddity.

Reviewer 14: doesn't an extra level of " " fix it?

2. `m.video` (from [MSC3553](https://github.com/matrix-org/matrix-doc/pull/3553))
3. `m.audio` / `m.voice` (from [MSC3246](https://github.com/matrix-org/matrix-doc/pull/3246)
   and [MSC3245](https://github.com/matrix-org/matrix-doc/pull/3245))
4. `m.image` (from [MSC3552](https://github.com/matrix-org/matrix-doc/pull/3552))
5. `m.file` (from [MSC3551](https://github.com/matrix-org/matrix-doc/pull/3551))

> 6. `m.location` (from [MSC3488](https://github.com/matrix-org/matrix-doc/pull/3488))

Reviewer 14: Am I to understand that these are defined as things that have meaning within a `content` but not within a `type`?

Reviewer 2: These are primary event types which will show up in `content`, yes.

> 9. `m.message` (implying `m.text` and `m.html` too)

Reviewer 14: I feel like having this list way earlier would have saved a whole lot of confusion on my part as I was reading this MSC.

This schema forms the minimum a client must emit for that event type, hwoever clients are welcome to
include additional types if they feel it is relevant. For example, a hypothetical IoT event added to the
spec might be as follows:

```json5
{
    "type": "m.temperature",
    "content": {
        "m.temperature": {
            "celsius": 27
        },
        "m.text": "it is 27 degrees"
    }
    // irrelevant fields excluded
}
```

In this example, the `m.temperature` event in `content` must be accompanied by an `m.message` fallback of
some kind, which can include a shortcut of `m.text` per this proposal.

Though required by the event schema, clients should not explicitly rely on other clients sending all of the
appropriate event types in `content`. If they are aware of what `m.temperature` is (for example), they will
not need the `m.message` components - clients which are unaware simply won't render the event, however.

In the case an event needs to fallback to `m.emote` or `m.notice`, the appropriate type can be included
in the event content as such:

```json5
{
    "type": "org.example.custom.event",
    "content": {
        "m.text": "is a fish",

>         "m.emote": {}, // denotes that this should fall back to an emote, but worst case gets rendered as text

Reviewer 15: What if someone sends the event:

```json
{
  "type": "org.example.custom.event",
  "content": {
    "m.text": "is a fish",
    "m.emote": {},
    "m.notice": {}
  }
}
```

Should it get rendered as an emote or a notice?  Or is this a case of "if you send such a silly event, the client gets to do whatever it wants with your event".

I'm somewhat concerned that the fact that the shortforms are used for three different primary types introduces a bunch of complexity and makes it likely that a client will do something wrong.  My understanding of the rules is:
- if the primary type is `m.message` (and the event does not already have an `m.message` property), then treat `m.text`/`m.html` as a shortform for `m.message`
- if the primary type is `m.emote`/`m.notice` (and the event does not already have a property of the same name), then treat `m.text`/`m.html` as a shortform for `m.emote`/`m.notice`.
- if the primary type is something different, and
  - there is a non-empty `m.emote`/`m.notice`/`m.message`, then use that, and ignore any `m.text`/`m.html` shortforms.  (Not sure what to do if there is, say, all three.)
  - there is an empty `m.emote`/`m.notice`, then use `m.text`/`m.html`, if present, as shortform for that type
  - otherwise, use `m.text`/`m.html` as shortform for `m.message`

Reviewer 2: it'd be client-defined behaviour. The rules early on say that with an unknown primary type a client should find the first representable event, which it can do in a number of ways.

Some clients might choose to grab the keys off the JSON and iterate through them until it's renderable (noting that objects in JSON are not stably ordered) while other clients might deliberately look for `m.emote` before `m.notice` and render as such.

Reviewer 13: Isn't the json always alphabetically ordered, because it needs to be canonically encoded over federation?

Reviewer 2: The order of keys in JSON have no ordering, so the client is expected to maintain a list of event types it prefers to fall back to.

    }
}
```

### Transition

It's simply not feasible to change the entirety of the Matrix ecosystem to a whole new set of events
overnight, so this proposal includes a time-constrained transition period to encourage all client
implementations to update their support for this schema before the ecosystem starts to move forward with
sending the newly-proposed primary event types.

Upon being included in a **released** version of the specification, the following happens:
* `m.room.message` is deprecated **but still used**.
* Clients include the fallback support described below in their outgoing events.
* Clients prefer the new format in events which include it.

> * Clients support the new primary event types, but continue to send `m.room.message` instead.

Reviewer 15: I think it would help to add a sub-bullet point here saying when sending the `msgtype` will be the event's primary type.  It is sort of discussed a bit later on, but I think it would help to have it more explicit at this location.

Reviewer 2: The msgtype and primary type are meant to be different things (`m.message` is the primary type for an `m.text` msgtype), but yes: clarification on the `msgtype` looks to be lacking.

the visibility of history before this MSC's introduction. Some clients, like bridges and bots, are likely
able to drop their support for rendering `m.room.message` events as they would have already processed the
older events (or have no reason/desire to backfill in a room).

As a fallback, a client simply smashes `m.room.message` together with the relevant event contents:

```json5
{
    "type": "m.room.message",
    "content": {
        "msgtype": "m.text",
        "body": "Hello World",
        "format": "org.matrix.custom.html",
        "formatted_body": "<b>Hello</b> World",
        "m.text": "Hello World",
        "m.html": "<b>Hello</b> World"
    }
}
```
or
```json5
{
    "type": "m.room.message",
    "content": {
        "msgtype": "m.text",
        "body": "Hello World",
        "format": "org.matrix.custom.html",
        "formatted_body": "<b>Hello</b> World",
        "m.message": [
            {"mimetype": "text/plain", "body": "Hello world"},
            {"mimetype": "text/html", "body": "<b>Hello world</b>"}
        ]
    }
}
```

A client can infer the intended primary type from the `msgtype` and otherwise engage their fallback process

> to try and render unknown `msgtype`s.

Reviewer 16: Following this comment at https://github.com/ruma/ruma/issues/790#issuecomment-1046362311:

> fwiw, the intent of the transitional format is that you don't look at the msgtype but rather the other parts of the extensible event format to determine what kind of event it is. The msgtype is only used when the event does not have an extensible event representation (or when extensible events are not being considered).

Reading only the MSC, I actually understood the opposite. Another sentence explaining that would probably be welcome.

Reviewer 16: Another concern: if we are not supposed to look at the `msgtype` for events with fallback, how do we know whether the  message is an `m.message`, `m.notice` or `m.emote`?

Must we add an empty `m.notice` or `m.emote` to be consistent with `m.voice`?

        }
    }
}
```

## Potential issues

It's a bit ugly to not know whether a given key will take a string, object or array.

It's a bit arbitrary as to which fields are allowed lists of fallbacks (eg image thumbnails).

It's a bit ugly that you have to look over the keys of contents to see what types
are present, but better than duplicating this into an explicit `types` list within the
event content (on balance).

We're skipping over defining rules for which fallback combinations to display
(i.e. "display hints") for now; these can be added in a future MSC if needed.
[MSC1225](https://github.com/matrix-org/matrix-doc/issues/1225) contains a proposal for this.

Placing event types at the top level of `content` is a bit unfortunate, particularly
when mixing in `m.room.message` compatibility, though mixes nicely thanks to namespacing.
Potentially conflicting cases in the wild would be namespaced fields, which would get
translated as unknown event types, and thus skipped by the rendering machinery.

This proposal additionally does not consider how other functionality like Replies, Edits,
Push Notifications, etc work. These are explicitly deferred to future MSCs.

> 

Reviewer 15: I think we have yet another (yay!) notifications issue here, in that AFAICT pings depend on matching text within the `body` field, which will be absent in a non-transitional extensible event.  So we'll need another MSC to fix notifications for this.

Reviewer 2: This is tracked as https://github.com/matrix-org/matrix-doc/issues/3555

* [MSC3488 - Location data](https://github.com/matrix-org/matrix-doc/pull/3488)
* [MSC3381 - Polls](https://github.com/matrix-org/matrix-doc/pull/3381)
* [MSC3245 - Voice messages](https://github.com/matrix-org/matrix-doc/pull/3245)
* [MSC2192 - Inline widgets](https://github.com/matrix-org/matrix-doc/pull/2192)
* [MSC3765 - Rich text topics](https://github.com/matrix-org/matrix-doc/pull/3765)

**Note**: Readers might find [Andy's blog](https://www.artificialworlds.net/blog/2022/03/08/comparison-of-matrix-events-before-and-after-extensible-events/)
useful for understanding the problem space. Unfortunately, for those who need to
understand the changes to the protocol/specification, the best option is to read
this proposal.

## Proposal

In a new room version (why is described later in this proposal), events are declared
to be represented by their extensible form, as described by this MSC. `m.room.message`
is formally deprecated by this MSC, with removal from the specification happening as
part of a room version adopting the feature. Clients are expected to use extensible
events only in rooms versions which explicitly declare such support (in both unstable

> and stable settings).

Reviewer 2: later in the MSC we mention those exceptions - will clarify up here to expect that text.

An extensible event is made up of two critical parts: an event type and zero or more
content blocks. The event type defines which content blocks a receiver can expect,
and the content blocks carry the information needed to render the event (whether the
client understands the event type or not).

Content blocks are simply any top-level key in `content` on the event. They can have
any value type (that is also legal in an event generally: string, integer, etc), and
are namespaced using the
[Matrix conventions for namespacing](https://spec.matrix.org/v1.4/appendices/#common-namespaced-identifier-grammar).

Content blocks can be invented independent of event types and *should* be reusable
in nature. For example, this proposal introduces an `m.markup` content block which
can be reused by other event types to represent textual fallback.

When a client encounters an extensible event (any event sent in a supported room
version) that it does *not* understand, the client begins searching for a best match
based on event type schemas it *does* know. This may mean combining multiple different
content blocks to match a suitable schema, such as in the case of MSC3553 video events.

> It is left as a deliberate implementation detail for which schemas to try, and in what

Reviewer 2: this behaviour is not necessarily client-specific, though it could be argued servers are acting as clients when consuming these events.

I'd rather leave it open if it's not affecting overall understanding of the MSC.

> It is left as a deliberate implementation detail for which schemas to try, and in what

Reviewer 14: I was just trying to correct the grammar here tbh. Possibly "Which schemas to try, and in what order, is left as a deliberate implementation detail."

> It is left as a deliberate implementation detail for which schemas to try, and in what

Reviewer 2: ah, that works - will adopt via https://github.com/matrix-org/matrix-spec-proposals/pull/1767/files#r1073903152

case of `m.markup` in `m.message`, or they could be optional depending on the situation.
Of course, senders are welcome to send even more blocks which aren't specified in the
schema for an event type, however clients which understand that event type might not
consider them at all.

In `m.message`'s case, `m.markup` is the only required content block. The `m.markup`
block can be reused by other events to include a text-like markup for the event, such
as a text fallback for clients which do not understand how to render a custom event
type.

To reiterate, when a client encounters an unknown event type it first tries to see
if there's a set of content blocks present that it can associate with a known event
type. If it finds suitable content blocks, it parses the event as though the event
were of the known type. If it doesn't find anything useful, the event is left as
unrenderable, just as it likely would today.

It is left as a deliberate client implementation detail to decide which content blocks
to search for, and in what order/combination. For example, a client might decide that
although it understands how to parse an `org.example.type` event, it will deliberately

> not try to "cast" unknown events into that type.

Reviewer 2: have removed this paragraph - it's duplicating information.

(see also: [MSC3933](https://github.com/matrix-org/matrix-spec-proposals/pull/3933) and friends).

### Power levels

This MSC proposes no changes to how power levels interact with events: they are still
capable of restricting which users can send an event type. Though events might be rendered
as a different logical type (ie: unknown event being rendered as a message), this does not
materially impact the room's ability to function. Thus, considerations for how to handle
power levels more intelligently are details left for a future MSC.

As of writing, most rooms fit into two categories: any event type is possible to send, or
specific cherry-picked event types are allowed (announcement rooms: reactions & redactions).
Extensible events don't materially change the situation implied by this power levels structure.

### Mixins specifically allowed

[Mixins](https://en.wikipedia.org/wiki/Mixin) are supported by Extensible Events to change
how an event is processed. All mixins are content blocks, and content blocks which are
mixins will be called out as such in the spec. Mixins are meant to be purely additive,

> thus all event types MUST support being rendered/processed *without* the use of mixins.

Reviewer 14: so to be clear: a mixin is a special sort of content block which exists only to be added on to other event types?

I feel like the boundary between what is, and is not, a mixin is pretty fuzzy. For example, MSC3551 defines an `m.caption`. It's optional, so the event supports being rendered/processed without it. What makes `m.automated` a mixin, but `m.caption` not?

Reviewer 2: it's a bit difficult to describe tbh. `m.caption` isn't a mixin because it can't universally be applied to any event type ("what does it mean to caption a poll"), which I guess might be the distinguishing factor?

Reviewer 2: I've added a note to this section to hopefully clear this specific concern up - let me know if it helps.

Reviewer 14: It helps, but I think it would be clearer to start out with a definition:

```suggestion
A **mixin** is a specific type of content block which can be added to any type of event to
change how that event is processed. Content blocks which are
mixins will be called out as such in the spec. Mixins are meant to be purely additive,
thus all event types MUST support being rendered/processed *without* the use of mixins.

See also the [Wikipedia entry on mixins](https://en.wikipedia.org/wiki/Mixin).
```

> Though this MSC does not describe any such mixins itself, MSC3955 does by allowing any

Reviewer 14: ```suggestion
Though this MSC does not describe any such mixins itself, [MSC3955](https://github.com/matrix-org/matrix-spec-proposals/pull/3955) does by allowing any
```

content blocks. The event type defines which content blocks a receiver can expect,
and the content blocks carry the information needed to render the event (whether the
client understands the event type or not).

Content blocks are simply any top-level key in `content` on the event. They can have
any value type (that is also legal in an event generally: string, integer, etc), and
are namespaced using the
[Matrix conventions for namespacing](https://spec.matrix.org/v1.4/appendices/#common-namespaced-identifier-grammar).

Content blocks can be invented independent of event types and *should* be reusable
in nature. For example, this proposal introduces an `m.markup` content block which
can be reused by other event types to represent textual fallback.

When a client encounters an extensible event (any event sent in a supported room
version) that it does *not* understand, the client begins searching for a best match
based on event type schemas it *does* know. This may mean combining multiple different
content blocks to match a suitable schema, such as in the case of
[MSC3553](https://github.com/matrix-org/matrix-doc/pull/3553) video events.

> It is left as a deliberate implementation detail for which schemas to try, and in what
> order. A client might decide to try parsing the event as a video, then image, then file,
> then text message, for example.

Reviewer 2: ```suggestion
Which schemas to try, and in what order, is left as a deliberate implementation detail.
A client might decide to try parsing the event as a video, then image, then file, then 
text message, for example.
```

ref https://github.com/matrix-org/matrix-spec-proposals/pull/1767/files#r1073357017

text message, for example.

It is generally not expected that a single content block will describe an entire event,
except in the exceedingly trivial cases (like text messages in this proposal). Multiple
content blocks will usually fully describe the information in the event, and mixins
(described later) can further change how an event is represented or processed.

Note that a "client" in an extensible events sense will typically mean an application
using the Client-Server API, however in reality a client will be anything which needs
to parse and understand event contents (servers for some functions like push rules,
application services, etc).

Per the introduction, text is the baseline format that most/all Matrix clients support
today, often through use of HTML and `m.room.message`. Instead of using `m.room.message`
to represent this content, clients would instead use an `m.message` event with, at
a minimum, a `m.markup` content block:

```json5
{
    // irrelevant fields not shown
    "type": "m.message",
    "content": {
        "m.markup": [
            { "body": "<i>Hello world</i>", "mimetype": "text/html" },
            { "body": "Hello world" }
        ]
    }
}
```

> `m.markup` has the following definitions associated with it:

Reviewer 2: `m.markup` rename thread

Reviewer 2: `m.text` feels more natural, though could lead to confusion when it's non-plain text?

Reviewer 2: `m.mime_text` is descriptive, though sounds weird. Doesn't feel like it covers future extensions like translations very well either.

Reviewer 17: `m.text_block` (it really should be a text block from the layout perspective)

Reviewer 17: `m.text_alternatives` (looong but conveys the idea that the array lists alternative content representations, not a sequence of those).

Reviewer 2: from @bleonard252 via email:

> How about *m.default*? Considering it would be present in almost every message, usable on its own, and as a fallback. Otherwise I second *m.text*, as it's concise and signifies it isn't intended for use for, say, images (unless it is?)

`m.default` feels a bit too out of the area as not all events will use it, by design. Internally we're rallying behind `m.text` though because it provides enough sensible meaning and any confusion would be clarified by another client spitting out parse errors.

Reviewer 2: `m.text` it is unless others have a better option.

Reviewer 17: @mscbot resolve `m.markup` is a weird name for a content block that will not have any markup 80% of the time

Reviewer 2: 71611964105311da79416c1d69b1cd16dff3e9a3

Author: oh, thank god this got renamed - thank you! :)

value type instead of the previously demonstrated objects and values - this is a legal
content block, as blocks can be of any type.

Clients should be cautious and avoid reusing too many unspecified types as it can create
opportunities for confusion and inconsistency. There should always be an effort to get
useful event types into the Matrix spec for others to benefit from.

### Room version

This MSC requires a room version to make the transition process clear and coordinated.
Normally for a feature such as this, an effort would be made to attempt to support
backwards compatibility for a duration of time, however for a feature that requires
significant overhaul of clients, servers, and Matrix as a whole it feels more important
to bias towards a clear switch between legacy and modern (extensible) events.

**Note**: A previous draft of this proposal (codenamed "v1 extensible events") did attempt
to describe a timeline-based approach, allowing for event types to mix concepts of content
blocks and legacy fields, however that approach did not give sufficient reason for clients
to fully adopt the extensible events changes.

> In room versions supporting extensible events, clients MUST only send extensible events.

Reviewer 18: Upgrading a room creates an entirely new room, "old" events will not be present in there.

> In room versions supporting extensible events, clients MUST only send extensible events.

Reviewer 19: noob alert!

Oh I wasn't aware of this, where does the spec mention it?

> In room versions supporting extensible events, clients MUST only send extensible events.

Reviewer 18: Here for example: https://spec.matrix.org/v1.8/client-server-api/#server-behaviour-16

> In room versions supporting extensible events, clients MUST only send extensible events.

Reviewer 19: Super thanks! I naively thought such an important thing would be mentioned in https://spec.matrix.org/v1.8/rooms/ . I still have to get used at how the spec chapters are organized.

the minimum scope of extensible events in a published (stable) room version.

Meanwhile, clients are welcome to use the unstable implementations of extensible event-supporting
features, provided they are in an appropriate room version. Some event type MSCs declare
explicit support for what would normally be an unsupported room version - client authors
should check the applicable MSC or specification for the feature to determine if they are
allowed to do this. Such examples include MSC3381 Polls and MSC3245 Voice Messages.

### State events

Unknown state event types generally should not be parsed by clients. This is to prevent situations
where the sender masks a state change as some other, non-state, event. For example, even
if a state event has an `m.markup` content block, it should not be treated as a room message.

Note that state events MUST still make use of content blocks in applicable room versions, and that
any top-level key in `content` is defined as a content block under this proposal. As such, this
MSC implicitly promotes all existing content fields of `m.*` state events to independent content

> blocks as needed. Other MSCs may override this decision on a per-event type basis (ie: redeclaring

Reviewer 2: 98bc9addba9fc0b5fcf6d214cb6018f223e67aac

change how that event is processed. Content blocks which are
mixins will be called out as such in the spec. Mixins are meant to be purely additive,
thus all event types MUST support being rendered/processed *without* the use of mixins.

See also the [Wikipedia entry on mixins](https://en.wikipedia.org/wiki/Mixin).

Note that mixins differ from optional content blocks in an event type's schema: a mixin
is able to be applied to *any* event type sensibly while optional content blocks are
generally only valuable to the applicable event types.

Though this MSC does not describe any such mixins itself,
[MSC3955](https://github.com/matrix-org/matrix-spec-proposals/pull/3955) does by allowing any
event to be flagged as "automated" - a strictly additive annotation on events.

Another possible mixin would be `m.relates_to` (not described by this MSC). Currently,
some features like the [key verification framework](https://spec.matrix.org/v1.5/client-server-api/#key-verification-framework)
rely on relationships as part of making the feature work. The expectation is that
these features would be adapted to meet the "purely additive" condition (assuming

> `m.relates_to` does actually end up being a mixin).

Reviewer 20: What about reactions those? Surely those can't function without a relation being present either.

Reviewer 2: yea... reactions are awkward. The relationship should be additive (without aggregation key), but representing the aggregation key as a content block is difficult. The expectation is that a dedicated MSC would solve extev reactions.

a [mention](https://spec.matrix.org/v1.5/client-server-api/#user-and-room-mentions) will
now appear under `m.markup`'s `text/html` representation (adding one if required).

A similar condition is applied to `body` in `m.room.message`: all existing functionality
will instead use the `text/plain` representation within `m.markup`, if not explicitly
called out by this MSC.

## Potential issues

It's a bit ugly to not know whether a given key in `content` will take a string, object,
boolean, integer, or array.

It's a bit ugly to not know at a glance if a content block is a mixin or not.

It's a bit ugly that you have to look over the keys of contents to see what blocks
are present, but better than duplicating this into an explicit `blocks` list within the
event content (on balance).

We're skipping over defining rules for which fallback combinations to display
(i.e. "display hints") for now; these can be added in a future MSC if needed.
[MSC1225](https://github.com/matrix-org/matrix-doc/issues/1225) contains a proposal for this.

Placing content blocks at the top level of `content` is a bit unfortunate, particularly

> when mixing in `m.room.message` compatibility, though mixes nicely thanks to namespacing.

Reviewer 2: err, no, it's not - will remove/reword (thought I did, honestly)

> when mixing in `m.room.message` compatibility, though mixes nicely thanks to namespacing.

Reviewer 2: Fixed in c5db1a059d4f62b4e8bba9abb545afa03c982f59

nicely thanks to namespacing. Potentially conflicting cases in the wild would be
namespaced fields, which would get translated as unrenderable events if the value type
doesn't meet the client's known schema.

This MSC does not rewrite or redefine all possible events in the specification: this is
deliberately left as an exercise for several future MSCs.

## Security considerations

Like today, it's possible to have the different representations of an event not match,
thus introducing a potential for malicious payloads (text-only clients seeing something
different to HTML-friendly ones). Clients could try to do similarity comparisons, though
this is complicated with features like HTML and arbitrary custom markup (markdown, etc)
showing up in the plaintext or in tertiary formats on the events. Historically, room
moderators have been pretty good about removing these malicious senders from their rooms
when other users point out (quite quickly) that the event is appearing funky to them.

> 

Reviewer 21: Power levels provide no provision for restricting what media (via mixins) you can embed into events.

Reviewer 20: @Gnuxie could you write this up as a new issue to make sure it doesn't get lost?

Reviewer 21: https://github.com/matrix-org/matrix-spec/issues/1932

shorthand for plain-text and html-formatted fallback for events. We lose `format`
and `msgtype`, however.

The various types in an event's contents MUST refer to the same event.  Multiple
events should be linked using a `m.relates_to` reference rather than multiplexing
into a single event.

Here are proposals for different types of events:

### m.text and m.html

For compactness for this super-common case, we define a short form which is
equivalent to the longer form `m.message` example below:

```json5
{
    "type": "m.message",
    "content": {
        "m.text": "i am a *fish*", // doesn't have to be markdown, but useful as a "fallback" for HTML
        "m.html": "i am a <b>fish</b>"
    }
}
```

### m.message

m.message describes a simple textual instant message.

```json5
{
    "type": "m.message",
    "content": {
        "m.message": [
            {
                "mimetype": "text/html", // optional, default text/plain
                "body": "i am a <b>fish</b>"
            },
            {
                "mimetype": "text/plain",
                "body": "i am a *fish*"
            }
        ],
    }
}
```

> ### m.file

Reviewer 2: to consider: what does encryption look like?

            {
                "url": "mxc://matrix.org/thumb",
                "mimetype": "image/jpeg",
                "width": 160,
                "height": 120,
                "size": 123
            }
        ],
    }
}
```

The schemas for some of the types above have some overlap, such as `m.image` and `m.thumbnail` repeating
the same information (or `m.file` also duplicating that information). De-duplicating this doesn't appear
to be an easy feat when we consider schema complexity, thus the duplication is tolerable by this MSC.

It's also worth noting that thumbnails do not have dedicated captions as they are implicitly a thumbnail
of the captioned content in the event. Events are meant to have exactly one thing represented within them,
meaning that a second caption is not required.

### m.message containing original source text, to allow future edits:

***TODO: Compare with how edits work in practice, adjust as needed.***

```json5
{
    "type": "m.message",
    "content": {
        "m.message": [
            {
                "mimetype": "text/html",
                "body": "I am a <a href=\"http://www.reddwarf.co.uk\">fish</a>",
            },
            {

>                 "mimetype": "text/markdown",

Reviewer: ```suggestion
                "mimetype": "text/markdown+commonmark",
```

Maybe it's useful to differentiate between the different types of markdown, to avoid conflicts in the future, were we ever to switch

Reviewer 6: I wonder if it is best for clients to convert the HTML to markdown so that it works for whatever variant they use. I don't know if we want markdown to be part of the protocol even if it is listed what type of even it uses. On the other hand when editing your own message in your own client you probably want to see the original markdown. If we are going to make it part of the protocol I would recommend we just pick one.

Reviewer 2: I'm ripping out all markdown references because it doesn't work properly within the confines of the spec.

this will require an amount of advertising and warning to land.

While this MSC is not considered stable by the specification, implementations *must* use
`org.matrix.msc1767` as a prefix to denote the unstable functionality. For example, sending
a `m.text` event would mean sending a `org.matrix.msc1767.text` event instead.

Currently visible events are `m.room.message` with a `msgtype` to distinguish
between `m.text`, `m.image`, `m.file`, etc. This MSC outlines how those would be 
translated to the new event shape, but does not cover fallback as obviously.

A suggested migration would be to send/receive `m.room.message` events as formatted today
with the additional fields from this MSC. For example, here's a simple text message:

```json5
{
    "type": "m.room.message",
    "content": {
        "msgtype": "m.text",
        "body": "Hello World",
        "format": "org.matrix.custom.html",
        "formatted_body": "<b>Hello</b> World",
        "m.text": "Hello World",
        "m.html": "<b>Hello</b> World"
    }
}
```

This looks very awful, but it at least shows that the parent type is meant to be a `m.text` event.

> Similar semantics apply to videos, audio, etc. This has a risk of potentially running up against the

Reviewer 11: #3052 defines how to include the waveform, that can be reused if wanted.

and the content blocks carry the information needed to render the event (whether the
client understands the event type or not).

Content blocks are simply any top-level key in `content` on the event. They can have
any value type (that is also legal in an event generally: string, integer, etc), and
are namespaced using the
[Matrix conventions for namespacing](https://spec.matrix.org/v1.4/appendices/#common-namespaced-identifier-grammar).

Content blocks can be invented independent of event types and *should* be reusable
in nature. For example, this proposal introduces an `m.markup` content block which
can be reused by other event types to represent textual fallback.

When a client encounters an extensible event (any event sent in a supported room
version) that it does *not* understand, the client begins searching for a best match
based on event type schemas it *does* know. For example, a client might attempt to
parse the event as a video event first, then try parsing it as audio, then as an
image, then as a file, and finally as a plain text message. If all of the client's
options are exhausted, the event is unrenderable by that client (as it would be

> prior to extensible events even existing).

Reviewer 22: Much as this section is already fairly complex, it might help to explain that content blocks are not necessarily alternative representations and there may be no single content block that fully describes the whole event - you might need to combine multiple content blocks to get the full picture, and call out to mixins.

Reviewer 2: ah, yes, that would be a good thing to call out. I've added some words to hopefully do so - let me know what you think.

Reviewer 2: Considering resolved due to FCP checkbox being ticked.

today, often through use of HTML and `m.room.message`. Instead of using `m.room.message`
to represent this content, clients would instead use an `m.message` event with, at
a minimum, a `m.markup` content block:

```json5
{
    // irrelevant fields not shown
    "type": "m.message",
    "content": {
        "m.markup": [
            { "body": "<i>Hello world</i>", "mimetype": "text/html" },
            { "body": "Hello world" }
        ]
    }
}
```

`m.markup` has the following definitions associated with it:
* An ordered array of mimetypes and applicable string content to represent a single
  marked-up blob of text. Each element is known as a representation.
* `body` in a representation is required, and must be a string.
* `mimetype` is optional in a representation, and defaults to `text/plain`.
* Zero representations are permitted, however senders should aim to always specify
  at least one.
* Invalid representations are skipped by clients (missing `body`, not an object, etc).
* The first representation a renderer understands should be used.
* Senders are strongly encouraged to always include a plaintext representation,

>   however are equally permitted to send an HTML representation instead.

Reviewer 2: I've dropped the bit after the comma - the first part is the important part ("please always send plain text at a minimum")

* The `mimetype` of a representation determines its `body` - no effort is made to
  limit what is allowed in the `body` (ie: we don't specify what is allowable HTML),
  therefore it is the client's responsibility to validate/sanitize the content further,
  such as by using the [existing spec](https://spec.matrix.org/v1.4/client-server-api/#mroommessage-msgtypes)

>   for recommended allowable HTML.

Reviewer 22: Well, before there is an explicit 'strongly suggested' list of tags, but here we're just mentioning it as one way clients might choose to do it. If we want to keep it the same, we should actually make the same suggestion. This could just be by saying, "the same set of strongly suggested tags from https://spec.matrix.org/v1.4/client-server-api/#mroommessage-msgtypes apply here too" or something.

>   for recommended allowable HTML.

Reviewer 2: Fixed in 68612194069ccf6a98655dc58335931452a0e0a4

to describe a timeline-based approach, allowing for event types to mix concepts of content
blocks and legacy fields, however that approach did not give sufficient reason for clients
to fully adopt the extensible events changes.

In room versions supporting extensible events, clients MUST only send extensible events.
Deprecated event types (to be enumerated at the time of making the room version) MUST NOT
be sent into extensible event-supporting room versions, and clients MUST treat deprecated
event types as unrenderable by force. For example, if a client sees an `m.room.message` in
an extensible event-supporting room version, it must not render it, even if it knows how
to render that type.

While full enforcement of this restriction is not feasible, servers are encouraged to block
Client-Server API requests for sending known-banned event types into applicable rooms. This
obviously does not help when the room is encrypted, or the client is sending custom events
in a non-extensible form, hence the requirement that clients treat the events as invalid too.

The Spec Core Team (SCT) will be responsible for determining when it is reasonable to include

> extensible events in a published (stable) room version.

Reviewer 2: It'll still be the regular MSC process (we'll need an MSC to make the stable room version happen anyways) - this is just clarifying that the SCT is responsible for determining which of the dozens of MSCs end up landing "first" rather than letting it evolve organically.

> Meanwhile, clients can send mixed support for extensible events if they prefer (such as by

Reviewer 2: I'm just going to remove this feature - it doesn't serve value.

### Power levels

This MSC proposes no changes to how power levels interact with events: they are still
capable of restricting which users can send an event type. Though events might be rendered
as a different logical type (ie: unknown event being rendered as a message), this does not
materially impact the room's ability to function. Thus, considerations for how to handle
power levels more intelligently are details left for a future MSC.

As of writing, most rooms fit into two categories: any event type is possible to send, or
specific cherry-picked event types are allowed (announcement rooms: reactions & redactions).
Extensible events don't materially change the situation implied by this power levels structure.

### Mixins specifically allowed

[Mixins](https://en.wikipedia.org/wiki/Mixin) are supported by Extensible Events through
dedicated content blocks. Though this MSC does not describe any such mixins itself,
MSC3955 does by allowing events to be flagged as "automated". Previously it was only possible
to flag a text message as an `m.notice` event, however with MSC3955 any event could become

> described as sent by a bot, thus reducing feedback loops.

Reviewer 2: I've added a couple paragraphs which I think might help, but might also be more noise - let me know what you think 😇

> described as sent by a bot, thus reducing feedback loops.

Reviewer 2: Considering resolved due to FCP checkbox being ticked.

to send a standalone `m.caption` event (for example) - this is not the case in this
iteration. Instead, an `m.caption` content block can be defined *without* specifying
an `m.caption` event type. However, event types and content block types can be named
the same way: they must not be interpretted to be the same logical type.

Per the introduction, text is the baseline format that most/all Matrix clients support
today, often through use of HTML and `m.room.message`. Instead of using `m.room.message`
to represent this content, clients would instead use an `m.message` event with, at
a minimum, a `m.markup` content block:

```json5
{
    // irrelevant fields not shown
    "type": "m.message",
    "content": {
        "m.markup": [
            { "body": "Hello world" },
            { "body": "<i>Hello world</i>", "mimetype": "text/html" }
        ]
    }
}
```

`m.markup` has the following definitions associated with it:
* An *unordered* array of mimetypes and applicable string content to represent a single
  marked-up blob of text. Each element is known as a representation.
* `body` in a representation is required, and must be a string.

> * `mimetype` is optional in a representation, and defaults to `text/plain`.

Reviewer 23: Is the intention to allow multiple bodies with the same mimetype?

If so, how should a client choose which one to display?  This smells like a vulnerability where a maliciously-crafted message could be made to display differently to different clients, e.g. Element shows "Alice is a wonderful person" while FluffyChat shows "Alice is a terrible person".

If the intent is not to allow multiple bodies with the same mimetype, then you could remove the possibility for confusion by making `content` a dictionary, with the mimetype as the key.

Reviewer: +1 for making `m.markup` a dictionary of mimetype -> body, should make things a lot clearer and implicit, unless the intention was to expand the body-mimetype object with other fields in the future.

Reviewer 24: Translation comes to mind. In theory this could still be handled with a subkey.

Reviewer 2: Translations, other markup details, extensions, etc - a dictionary is not fit for purpose here, sorry. Seeing two different messages is already an issue, and this MSC doesn't make it better or worse.

* Zero representations are permitted, however senders should aim to always specify
  at least one.
* Invalid representations are skipped by clients (missing `body`, not an object, etc).
* The client can choose which representation fits its purpose best, though generally

>   the preferred representations should appear first in the array.

Reviewer: This does not feel "unordered", it feels there are counteractive signals here;
- "this is an unordered array"
- "the preferred representation *could* be ordered first"

----

For example, if (like the above example) a text/plain markup is ordered before text/html, which one would be "preferred"? Does that mean that the client should prefer the text/plain repr over the text/html one?

----

It also seems that [this section](https://github.com/matrix-org/matrix-spec-proposals/pull/1767/files#diff-242c41793e638a8acaf2b7ea5870ca0249de76f00b36e8fb2c2e5ed4e106872aR125-R129) implies the reinforce the first note, while disregarding the second one.

----

I believe this section could be worded better in terms of fallback, or not present at all.

Reviewer 2: Originally this field was ordered, but that caused problems because people naturally put stuff like plain text first, then HTML, then a lot of the more optional formats, leading to a problem where the second array element is actually "first prepared", then the first element, then the remainder. Instead, we are trying to tell clients to pick the one they want to use (which they should be able to do safely anyways), but as a *suggestion*, pick the first one if you don't have a good option.

Reviewer: That sounds reasonable, though I just think it's not communicated clearly in the proposal here.

Reviewer 2: > That sounds reasonable, though I just think it's not communicated clearly in the proposal here.

will attempt to clarify - it was a last minute change (per commits).

> On that note: should events building upon ext ev suggest an order? As in: "its fine to use whatever but for best compatibility clients should prefer this order if they support it."
> 
> Meaning that if a client supports variant a and b and the event sends a b and c as possible representations _should_ the client prefer a over b or b over a for an event to make sure that clients are consistent across the ecosystem if the client supports both?
> 
> While for HTML vs unformmated messages is very clear what the best way of order is when a client supports both it may not be as clear with let's say for example having an imaginary ext ev where you have 2 image formats to choose from. (this is just an example to display the issue). As in for ext evs where there are representations that can have equal weight.

The client would know better than the sender for what representations are best for it to render, especially when considering extensions to the markup system (translations, etc). Leaving it as a deliberate implementation detail seems wise, imo.

Reviewer 2: actually, given all the other comments, I think we should just go back to making this ordered. This would also apply to notifications stability and such, which feels important.

* Senders are strongly encouraged to always include a plaintext representation,

>   however are equally permitted to send an HTML representation instead.

Reviewer: ```suggestion
  however are equally permitted to only send an HTML representation instead.
```
?

Reviewer 2: The lack of "only" is intentional: senders are asked to send plain text with their Markdown, but they can also send HTML+Markdown instead.

* The `mimetype` of a representation determines its `body` - no effort is made to
  limit what is allowed in the `body` (ie: we don't specify what is allowable HTML),

>   therefore it is the client's responsibility to validate/sanitize the content further.

Reviewer: This feels untrue, as there is a [section in the spec](https://spec.matrix.org/v1.4/client-server-api/#mroommessage-msgtypes) that details which html tags and attributes are allowed and which aren't. Do those get grandfathered into here, or does this effectively release those restrictions?

Personally, I would advice against releasing the restrictions, as that will introduce vagueness and uncertainty around how html will be rendered on different clients. The existing set of HTML tags and attributes help sanitise and "bound" what to expect from the html over the wire.

Reviewer 2: Those restrictions are already a suggestion - this MSC doesn't neccessarily remove that section of the spec, but it also doesn't really feel constrained by it. Clients can (and should) still sanitize as needed, including following recommendations in the spec about how HTML can work.

For the features of our HTML that aren't allowed/expected in "normal" HTML, we should open an MSC for a MatrixHTML specification.

* Custom markups in a representation are specified by a suitably custom `mimetype`.
  For example, extending HTML or XML or an all-new format. This can be used to create
  bridge-compatible clients where the destination network's markup is first in the

>   array, followed by more common HTML and text formats.

Reviewer: I'd like to make a note here (not a review comment) about that this would allow the likes of [`text/x-gfm`](https://www.digipres.org/formats/mime-types/#text/x-gfm) to be a proper mimetype to target, for markdown-compatibility-related purposes.

If possible, it might be interesting to include this as a sub-item, to evoke the imagination or have readers have an "ah" moment about this, as I know markdown compatibility has been somewhat of a sore point, and so this could be used to "lump" the raw markdown into the message, while keeping compatible rich representation in the markdown.

Reviewer 2: I personally feel the MSC is already too long and markdown examples don't add to the vision the MSC is trying to portray - instead, it's mentioned as a "oh look, markdown too" to remind readers there are more mimetypes than text and HTML. Specifically, this is actually targeting Slack, Wire (to a degree), and other messengers which use highly custom formats.

value type instead of the previously demonstrated objects and values - this is a legal
content block, as blocks can be of any type.

Clients should be cautious and avoid reusing too many unspecified types as it can create
opportunities for confusion and inconsistency. There should always be an effort to get
useful event types into the Matrix spec for others to benefit from.

### Emotes and notices

Though regular text messages cover a lot of cases, `m.room.message` also supports emote
and notice as possible message types. Under extensible events, the `m.emote` event type
replaces the `m.emote` msgtype, as does `m.notice` for notice events, as shown:

```json5
{
    // irrelevant fields not shown
    "type": "m.emote",
    "content": {
        "m.markup": [{"body": "says hi"}]
    }
}
```

```json5
{
    // irrelevant fields not shown
    "type": "m.notice",
    "content": {
        "m.markup": [{"body": "this is a bot message"}]
    }
}
```

These event types do not define any additional content blocks, instead reusing the markup
block from earlier in this proposal. `m.markup` is a required block for these event types.

> To support event types which may need an emote or notice fallback, `m.emote` and `m.notice`

Reviewer 25: Can you explain, why this is needed? Is there an actual use case?

In my opinion we should should distinguish between event type (e.g. `m.message` or `m.emote`) and content schemas (e.g. `m.markup` or `org.exmaple.probe_value`) to avoid confusion. This paragraph does mix them up. I also think, that it overcomplicates de-/serializersation

Reviewer 2: `m.notice` is more useful of the two, as (hopefully) demonstrated by the later example. `m.emote` also gets a content block definition because, well, we're here and it's easy to do so, and costs us nothing. Theoretically, one could use `m.emote` as a fallback for `m.sticker` or similar, to communicate rich-enough integration.

As for how it affects (de)serialization, there shouldn't be much of an impact: content blocks are specifically called out in the MSC as *not* being event types, though they can have the same name. Some effort was made to avoid `m.message` and `m.message` naming (instead using `m.markup` for the content block), however for notices and emotes it's hard to avoid the terminology overlap. A parser/serializer would treat them as completely different things, even if named the same.

Reviewer: I believe embedded `m.notice` has a solid use-case here, but `m.emote` seems questionable as a candidate for being embeddable.

Reviewer 2: > So there is a distinction between event type and content blocks?

Yes, explicitly (see early paragraphs in the proposal).

> > m.emote and m.notice are defined as content blocks with an object containing the content of a normal m.emote or m.notice event.
> 
> Still, this sounds to me like content block types can be nested, which has a compexity impact on de-/serialization and is really confusing. I want to understand why this decision was made.
> 
> As an alternative there could be an `m.emote` and `m.notice` content block boolean without nesting the `m.markup`. Thereby a client must not support `m.emote` and `m.notice` to render a content, because `m.markup` would stay on top level.

They can be nested, but they would be nested with explicit schema (so should be possible to parse still).

Reviewer 23: It is weird that `m.markup` is initially presented as the place to look for a fallback.  And now we have `m.notice` and `m.emote` also available as different fallback options.  But then they're just containers for the same structure that we had before.

I guess what I'm asking is: What is the reason for having the extra layer of `m.notice` here?  Why not just use the `m.markup` directly under `content` like you described above?

Would some event type want to have both an `m.markup` and an `m.notice`, and have them say different things?

Reviewer 25: I want to point out my alternative again:

> As an alternative there could be an m.emote and m.notice content block boolean without nesting the m.markup. Thereby a client must not support m.emote and m.notice to render a content, because m.markup would stay on top level.

Reviewer 12: I think the underlying trouble is that the MSC *is* mixing up event types and content blocks semantically, to a degree. What is a `m.notice` content block even supposed to mean? I suppose it wants to say "my inner contents represent a notice, rather than some markup which *isn't* a notice".

But that sounds squarely what the purpose of the `m.notice` *event type* is supposed to be. So why are we trying to nest it as a content block inside some other type?

Reviewer 2: @cvwright You can put `m.markup` one layer up, but then it gets rendered as a text message rather than as a notice. This is why it's nested.

@benkuly while I appreciate the suggestion, it's been dismissed before because a boolean is not extensible enough for future consideration, unfortunately. 

@dkasak The semantics are indeed being combined, deliberately. It's to say that the event gets rendered like a notice *if the client doesn't believe there's a more suitable content block*, which then means it also uses the text. This does not mean that the client renders it as a probe value (for example) that is faded because it's a notice - it would either render as a probe value or as a notice, not both.

Reviewer 12: Then the problem is that the `m.notice` event type has unclear semantics. What does the `m.notice` type mean, other than its faded rendering? Does it mean anything at all or is it just a styling thing?

Reviewer 12: To attempt to pinpoint what I think is wrong about this: It makes the answer to the question of "Is an event a notice?" hard and unobvious. Previously, you'd just check the event type and, if it's `m.notice`, it's a notice, and otherwise it's not a notice.

But with an `m.notice` content block, it *may* be a notice to some clients, but only if they don't know how to render the other content blocks? Or, alternatively, it's *not* a notice for anyone, it just renders like one on some clients.

This is the result of mixing semantics and presentation and not clearly delineating between them.

Reviewer 25: > @benkuly while I appreciate the suggestion, it's been dismissed before because a boolean is not extensible enough for future consideration, unfortunately.

Okay, then make it an object:

```json5
{
    // irrelevant fields not shown
    "type": "org.example.temperature",
    "content": {
        "m.markup": [{"body": "It is 22 degrees at Home"}],
        "m.notice": {
            "enabled": true
        },
        "m.emote": {
            "enabled": true
        },
        "org.exmaple.probe_value": {
            "label": "Home",
            "units": "org.exmaple.celsius",
            "value": 22
        }
    }
}
```

or include it in `m.markup`:

```json5
{
    // irrelevant fields not shown
    "type": "org.example.temperature",
    "content": {
        "m.markup": {
            "body": [{"body": "It is 22 degrees at Home"}],
            "m.notice": true,
            "m.emote": true
        }
        "org.exmaple.probe_value": {
            "label": "Home",
            "units": "org.exmaple.celsius",
            "value": 22
        }
    }
}
```

Reviewer 2: It has no semantic difference to `m.room.message` with `msgtype` of `m.notice` - we just aren't letting events subtype themselves with things like `msgtype` anymore.

Reviewer 12: But I think the problem is exactly that the event type *should* carry semantics. If as you say `m.notice` does *not* have different semantics than `m.message`, just different styling, then we should be sending notices as `m.message` and styling them differently by using an appropriate content block, such as the one suggested above:

```json
"m.markup": {
    "body": [{"body": "It is 22 degrees at Home"}],
    "m.notice": true
}
```

Reviewer 2: > To attempt to pinpoint what I think is wrong about this: It makes the answer to the question of "Is an event a notice?" hard and unobvious. Previously, you'd just check the event type and, if it's `m.notice`, it's a notice, and otherwise it's not a notice.
> 
> But with an `m.notice` content block, it _may_ be a notice to some clients, but only if they don't know how to render the other content blocks? Or, alternatively, it's _not_ a notice for anyone, it just renders like one on some clients.
> 
> This is the result of mixing semantics and presentation and not clearly delineating between them.

@dkasak If a client knows how to render the event type then the client renders it as such. If it *doesn't* know how to render that event type, then it starts looking at content blocks. The client picks which content blocks are looked at first, deliberately, and might in fact choose a different one from another client. The notice is effectively a text fallback if nothing else seems relevant to that client.

---

@benkuly That is closer to mixins, not independent rendering, which is what we're avoiding at the moment. Display hints and similar are explicitly out of scope.

Reviewer 12: @turt2live I understand the presentation side of this and how the rendering rules work. I'm rather concerned we're eroding event *meaning* (semantics) here, by mixing semantic concerns with presentational concerns. Each event type should have a clearly defined meaning without reference to the way it will be presented by clients.

Reviewer 2: From out of band, it sounds like we need to declare that mixins are specifically allowed, and likely push emotes and notices out to their own MSCs for further discussion. Specifically, around notices, it feels more desirable to have a mixin to say "this event is automated - do not reply" rather than an event type to do that.

Reviewer 2: Emotes are now https://github.com/matrix-org/matrix-spec-proposals/pull/3954 and notices are https://github.com/matrix-org/matrix-spec-proposals/pull/3955

The proposal text here is updated in 48e058849e6d47eb79e4ec6df26d6e1f7dc900fe

are defined as content blocks with an object containing the `content` of a normal `m.emote`
or `m.notice` event:

```json5
{
    // irrelevant fields not shown
    "type": "org.example.temperature",
    "content": {
        "m.notice": {
            "m.markup": [{"body": "It is 22 degrees at Home"}]
        },

>         "org.exmaple.probe_value": {

Reviewer 12: ```suggestion
        "org.example.probe_value": {
```

Reviewer 2: will apply manually

and clients MUST treat `m.room.message` as an invalid event type.

While enforcement of this restriction is not possible, servers are encouraged to block client-server
API requests for sending known-banned event types into applicable rooms. This obviously does not
help when the room is encrypted though, which is why clients MUST treat the deprecated event types
as invalid in the room.

The Spec Core Team (SCT) will be responsible for determining when it is reasonable to include extensible
events in a published room version.

Meanwhile, client authors can send mixed support for extensible events if they prefer, however SHOULD
NOT send largely-unknown event types into room versions which do not support extensible events. For
example, a client might send `m.room.message` events with `m.markup` content blocks, but should not
send the new `m.message` event type into that same room.

### State events

State events typically contain metadata about the room and are not rendered themselves, however there
are some cases where rendering the state event change might be useful. For example, it might be desirable

> to render a bridge state change with `m.markup` to clearly indicate a bridge was added/edited.

Reviewer 23: If you allow state events to include `m.markup`, how does a client validate that the textual description is accurate and not misleading?

For example, "Alice made Bob a moderator" vs "Alice muted Bob", or "Alice invited Bob" vs "Alice banned Bob"?

Reviewer 26: Also, how would a client here deal with translation for this? And/Or when a client uses a different scheme for their context messages? (We already have clients using different product names for features.)

Reviewer 2: There's several things here:

1. Clients are not required to render extensible events on state events.
2. Room admins/moderators that do this (because you need permission to send a state event) are unlikely to be well-received by the room's community.
3. The fallback is only used when the client doesn't know how to deal with a given state event, which means normal users (who can send membership events) can't actually do any damage: their membership event would be rendered as a membership event without any fallback. The same applies for the power levels event.
4. Even if the client *does* decide to render extensible events on state events, falling back to `m.markup` means it'll get rendered as a text message, not as a state event. Thus it'll be no different than literally saying "Alice made Bob a moderator" in chat.
5. On translations, please refer to the other MSC and point 4 above.

Reviewer 12: > Even if the client does decide to render extensible events on state events, falling back to m.markup means it'll get rendered as a text message, not as a state event. Thus it'll be no different than literally saying "Alice made Bob a moderator" in chat.

It would be *rendered* as a text message, but would it still have the *meaning* of a state event?

Reviewer 12: > > It would be rendered as a text message, but would it still have the meaning of a state event?
>
> Yes.

Well isn't this exactly the problem pointed out above? Because state events usually have effects, and so their presentation should reflect this effect, but if they get rendered as a text message, the message can say anything at all, even wording contrary to the effect?

Reviewer 2: it'd be no different than blatantly lying in a text message though? I'm not sure technology can solve this problem space, or if it should.

Reviewer 12: Surely it is different, because ordinary text messages don't make the client perform side-effects. The concern is that someone would be able to use this to perform a side-effect but then mask the fact of that happening through this mechanism.

Reviewer 12: From out of band, the case I'm most worried about is something like a client implementing a state event but continuing to let "fallback" content blocks to inform and override its rendering. For a concrete example, a client implements `org.example.door_lock`, but instead of only relying on the `door_state` content block (specified in its event schema) to inform the rendering, it allows itself to be overridden by a 1px `m.image` content block, rendering the state event invisible.

The solution is to add some wording to say that clients should not do this and to warn implementors of the vulnerability.

Reviewer 23: > > Can't we just say that standard state events (those beginning with `m.`) don't support content blocks to solve this problem?
> 
> Preferably not, because special casing causes problems. I also don't really see how this solves the problem?

Distinguishing state events vs non-state events is not really special casing though, is it?  They are two fundamentally different things in the spec.

Reviewer 23: > From out of band, the case I'm most worried about is something like a client implementing a state event but continuing to let "fallback" content blocks to inform and override its rendering. 

Yes exactly.  To put a somewhat finer point on it, suppose a compromised homeserver maliciously added a user to the room.  The story all along has been that "Users should pay attention to room membership".  But now that defense just got a lot more difficult, because now the `m.room.member` event in the timeline can be disguised with a fallback message / cover story.

Reviewer 2: We discussed that in the room: clients should be treating member events as known, but point taken that it's possible for developers to make mistakes. The requirements @dkasak talks about are actually already in the proposal, just not obviously.

Reviewer 24: Recognized state events, like m.room.member events in most clients, obviously wouldn't use m.markup. However, we could recommend clients to show a warning that "An unknown change was made."

Reviewer 2: As of 48e058849e6d47eb79e4ec6df26d6e1f7dc900fe this MSC no longer recommends state events to be treated as timeline events at all, instead doing the formality of promoting existing `content` bodies of state events to relevant content blocks (not that anyone is expected to deal with them - see updated proposal text).

State events MUST make use of content blocks, noting that values like `join_rule` and `membership` found
in `content` today are implicitly legal content blocks, however are not required to offer "sensible"
fallback options for rendering.

### Encryption

Like `m.room.message`, `m.room.encrypted` is also deprecated in favour of a new `m.encrypted` event
type. `m.encrypted` expects an `m.encrypted` content block, which is the current `content` schema for

> an `m.room.encrypted` event:

Reviewer: Reusing a name for a content block and a type seems to become confusing to some, I could recommend using `m.ciphertext` as a name for the content block, but I see that `ciphertext` is also used as a field under that content block, so I'd only strongly want to note the possibility of confusion between the two places where `m.encrypted` could be used.

```json5
{
    // irrelevant fields not shown
    "type": "m.encrypted",
    "content": {
        "m.encrypted": {
            "algorithm": "m.olm.v1.curve25519-aes-sha2",
            "sender_key": "<sender_curve25519_key>",
            "ciphertext": {
                "<device_curve25519_key>": {
                    "type": 0,
                    "body": "<encrypted_payload_base_64>"
                }
            }
        }
    }
}
```

> This allows the `m.encrypted` content block to be reused by other event types, if required.

Reviewer 12: Why would this be required? Are we already suspecting it or is this just an off-handed remark?

Reviewer 24: This seems like it would be intended to allow or encourage non-encrypted data to be sent along with encrypted data. As far as progress on encryption goes, this seems unhelpful at best.

Reviewer 23: Could this be used as an alternative to #3911 for linking media to encrypted events?

That is, the encrypted content stays the same as before, and you could add a list of MXC media URL's elsewhere under the `content`.

Reviewer 2: > > For the extensible events world though, this is to make "encryption" be a single content block instead of 3, as all event types get redefined over time.
> 
> Hmm, can you expand a bit? Which (hypothetical) 3 are you referring to?

They aren't hypothetical 😇 Any top-level key in `content` is a content block, so the three fields currently required to send an encrypted message each become a content block.

> Could this be used as an alternative to #3911 for linking media to encrypted events?
> 
> That is, the encrypted content stays the same as before, and you could add a list of MXC media URL's elsewhere under the `content`.

Unlikely - that would make things more complicated.

Reviewer 12: Ah, okay. To spell it out for others, I suppose you meant `algorithm`, `session_id` and `ciphertext`, and excluding `sender_key` and `session_id` because they are deprecated.

Reviewer 2: Special cases are really not ideal, as it causes so many more problems. I'd sooner specify that the `m.encrypted` content block should not be used to have encrypted/unencrypted fallbacks.

Note that it's already possible to send unencrypted content in an encrypted event - this MSC isn't really changing anything, but rather formally making encryption a thing you can represent more easily under this system.

Reviewer 23: > > Could this be used as an alternative to #3911 for linking media to encrypted events?
> > ...
> 
> Unlikely - that would make things more complicated.

Sorry, I meant that more as "Is this the kind of thing you are envisioning?" more than "Do you think this particular application is a good idea?"

Overall I don't see a serious problem here.  There are often legitimate reasons to send plaintext data alongside of some ciphertext.  Articulating such a use case here might help to justify this part of the proposal, but even without one it's not a stretch to believe that they could exist.

Clients just need to not be stupid about what they send in the clear, and as you said, this is already a requirement.

Reviewer 23: Actually on second thought, allowing encrypted content blocks in lots of different message types could be quite cool.

Then the server could filter messages by type, reducing the number of messages that a client must download and decrypt.

Reviewer 2: Encrypted events are now their own dedicated MSC for further discussion: https://github.com/matrix-org/matrix-spec-proposals/pull/3956

This proposal is updated as of 48e058849e6d47eb79e4ec6df26d6e1f7dc900fe to reflect this.

### Notifications

Currently [push notifications](https://spec.matrix.org/v1.3/client-server-api/#push-notifications)
describe how an event can cause a notification to the user, though it makes the assumption
that there are `m.room.message` events flying around to denote "messages" which can trigger
keyword/mention-style alerts. With extensible events, the same might not be possible as it
relies on understanding how/when the client will render the event to cause notifications.

> For simplicity, the `content.body` condition now looks for an `m.markup` block's `text/plain`

Reviewer 14: > the `content.body` condition

It's unclear to me what exactly this refers to. Note that there is no `content.body` condition defined at https://spec.matrix.org/v1.4/client-server-api/#conditions-1.

Perhaps you mean  `event_match` conditions with a key of `content.body`? It feels odd to me that we would special-case some keys to give them special behaviour, and really we should define a whole new condition type here. On the other hand, `content.body` already has some special-casing for historical reasons, so I suppose adding more special cases isn't the end of the world.

Either way, clarification needed here please.

Reviewer 14: Also: does this change to the behaviour apply for all room versions, or just ones that support extensible events?

Reviewer 2: This feels a bit like over-specification in an already very long MSC, imo. It feels clear enough to me that this is targeting `event_match` with `content.body`, and that all other improvements (because there are obviously some) are deferred. Other MSCs are already mentioned as being needed to handle the `m.room.message` -> `m.message` event type transition anyways - I'd rather move this discussion to said future MSC.

Reviewer 2: have clarified anyways, I think.

Reviewer 14: yeah that's clearer, thanks

applicable push rules to handle the new event types.

A future MSC is expected to address the default push rules having a hardcoded event type.

A future MSC is also expected to introduce a better system for handling push notifications
on extensible events. This MSC aims for basic compatibility with the shortest turnaround
rather than getting riled in discussion about a notifications system structure.

### Power levels

This MSC proposes no changes to how power levels interact with events: they are still
capable of restricting which users can send an event type. Though events might be rendered
as a different logical type (ie: unknown event being rendered as a message), this does not
materially impact the room's ability to function. Thus, considerations for how to handle
power levels more intelligently are details left for a future MSC.

As of writing, most rooms fit into two categories: any event type is possible to send, or
specific cherry-picked event types are allowed (announcement rooms: reactions & redactions).
Extensible events don't materially change the situation implied by this power levels structure.

## Potential issues

> It's a bit ugly to not know whether a given key in `content` will take a string, object or array.

Reviewer 23: I can't speak for the other languages with real type systems, but all of this looks very doable in Swift using Apple's `Decodable` protocol.  Much nicer than, say, the params for `m.login.terms`.  :-)

That is, as long as there's an unambiguous spec for each message type.

Reviewer 2: Good to hear! I'm going to resolve this thread though so I don't think there's something to action ;)

This MSC does not rewrite or redefine all possible events in the specification: this is
deliberately left as an exercise for several future MSCs.

## Security considerations

Like today, it's possible to have the different representations of an event not match,
thus introducing a potential for malicious payloads (text-only clients seeing something
different to HTML-friendly ones). Clients could try to do similarity comparisons, though
this is complicated with features like HTML and arbitrary custom markup (markdown, etc)
showing up in the plaintext or in tertiary formats on the events. Historically, room
moderators have been pretty good about removing these malicious senders from their rooms
when other users point out (quite quickly) that the event is appearing funky to them.

## Unstable prefix

While this MSC is not considered stable by the specification, implementations *must* use
`org.matrix.msc1767` as a prefix to denote the unstable functionality. For example, sending
an `m.message` event would mean sending an `org.matrix.msc1767.message` event instead.

Implementations wishing to trial the migration ahead of this MSC's inclusion in the spec would

> use the unstable prefix mentioned above. A fallen back `m.room.message` example event becomes:

Reviewer: (Forgotten during the rewrite?)

```json5
{
    "type": "m.room.message",
    "content": {
        "msgtype": "m.text",
        "body": "Hello World",
        "format": "org.matrix.custom.html",
        "formatted_body": "<b>Hello</b> World",

>         "org.matrix.msc1767.text": "Hello World",
>         "org.matrix.msc1767.html": "<b>Hello</b> World"

Reviewer: ```suggestion
        "org.matrix.msc1767.markup": [
            {"body": "Hello World"}, 
            {"body": "<b>Hello</b> World", "mimetype": "text/html"}
        ]
```

Reviewer 2: will apply manually - there's some concerns with this layout.

    }
}
```

For purposes of testing, implementations can use a dynamically-assigned unstable room version
`org.matrix.msc1767.<version>` to use extensible events within. For example, `org.matrix.msc1767.10`
for room version 10 or `org.matrix.msc1767.org.example.cool_ver` for a hypothetical
`org.example.cool_ver` room version. Any events sent in these room versions *can* use stable
identifiers given the entire room version itself is unstable, however senders *must* take care
to ensure stable identifiers do not leak out to other room versions - it may be simpler to not
send stable identifiers at all.

## Changes from MSC1225

* converted from googledoc to MD, and to be a single PR rather than split PR/Issue.
* simplifies it by removing displayhints (for now - deferred to a future MSC).
* removes all references to mixins, as the term was scaring people and making it feel far too type-theoretic.
* replaces the clunky m.text.1 idea with lists for types which support fallbacks.
* removes the concept of optional compact form for m.text by instead having m.text always in compact form.
* tries to accomodate most of the feedback on GH and Google Docs from MSC1225.

> ## Historical changes

Reviewer: Ditto, feels like https://github.com/matrix-org/matrix-spec-proposals/pull/1767#issuecomment-1310393397 should be included here.

Reviewer 2: It is? The parts highlighted for advertising on github/the spec room are intentionally different than the ones recorded in the MSC's own changelog. The MSC's changelog is meant to cover drastic changes whereas the "please take another look" comment highlights the subtlties to encourage readers with opinions to re-read it.

based on an earlier draft of this proposal (as of writing). Unfortunately, for
those who need to understand the changes to the protocol/specification, the only
option is to read this proposal.

## Proposal

In a new room version (why is described later in this proposal), events are declared
to be represented by their extensible form, as described by this MSC. `m.room.message`
is formally deprecated by this MSC, with "removal" happening as part of the room
version adoption. Clients are expected to use extensible events only in rooms versions
which explicitly declare such support.

An extensible event is made up of two critical parts: an event type and zero or more
content blocks. The event type describes the information being carried in the event,
and the content blocks describe how to render that information (whether the client
understands the event type or not).

Content blocks are simply any top-level key in `content` on the event. They can have
any value type, and are namespaced using the
[Matrix conventions for namespacing](https://spec.matrix.org/v1.4/appendices/#common-namespaced-identifier-grammar).

> Unlike previous iterations of this proposal, it is not expected that content block

Reviewer 27: This paragraph is useful for people familiar with history, and very confusing for someone reading this the first time. Can we make it a sidebar, or relegate it to a footnote?

Reviewer 2: Clarified in 48e058849e6d47eb79e4ec6df26d6e1f7dc900fe

    }
}
```

These event types do not define any additional content blocks, instead reusing the markup
block from earlier in this proposal. `m.markup` is a required block for these event types.

To support event types which may need an emote or notice fallback, `m.emote` and `m.notice`
are defined as content blocks with an object containing the `content` of a normal `m.emote`
or `m.notice` event:

```json5
{
    // irrelevant fields not shown
    "type": "org.example.temperature",
    "content": {
        "m.notice": {
            "m.markup": [{"body": "It is 22 degrees at Home"}]
        },
        "org.example.probe_value": {
            "label": "Home",
            "units": "org.example.celsius",
            "value": 22
        }
    }
}
```

While technically possible to include an `m.notice`/`m.emote` content block on an `m.message`
event, doing so wouldn't really do much: as most clients are expected to understand `m.message`
events, they would ignore `m.notice` or `m.emote` blocks. Adding the block does *not* turn the
event into a notice/emote - it is still a regular text message, and the sender probably should
have used the `m.notice` or `m.emote` event type specifically.

> ### Room version

Reviewer 2: How does the client know it is speaking in a suitable room version though? Hardcoding works for the first room version, but as soon as we add a second room version we'll have a bunch of outdated clients.

Probably shove something into the capabilities API?

Reviewer 28: MSC3244 attempted to handle this for spaces, but has since been rejected.

Reviewer 2: From memory, the issue with MSC3244 was that it was attempting to describe a server preference for room version features, not considering the client's need, and was for *creating* rooms. The difference here would be *using* rooms.

Reviewer 2: I'm implicitly deferring this to a future MSC to figure out. For now, hardcoding room versions which support/don't support extensible events is working well enough (but won't work in the later stages of the project: see spec process section of this MSC)

{
    "type": "m.emote",
    "content": {
        "m.message": [
            { "mimetype": "text/plain", "body": "is a fish" }
        ]
    }
}
```
```json5
{
    "type": "m.notice",
    "content": {
        "m.message": [
            { "mimetype": "text/plain", "body": "i am a fish that is sending a notice" }
        ]
    }
}
```

To use `m.emote` or `m.notice` as a reusable event type on a custom/other primary event type
the event type can be associated with an empty object:

```json5
{
    "type": "org.example.custom.event",
    "content": {
        "m.text": "is a fish",
        "m.emote": {}, // denotes that this should fall back to an emote, but worst case gets rendered as text
    }
}
```

**Note**: it is technically possible with this structure to have someone send an event that is
both a notice and emote. Given order of preference is a deliberate implementation detail, it is
left to the implementation to determine how it wants to handle this. It might, for example, show
the event as an emote with a bot-like style, or simply as a notice (no emote semantics).

#### Shortcuts

In order to make demonstrating the protocol significantly easier and to help developers/curious

> invididuals in understanding what a message event is, two shortcuts are defined: `m.text` and

Reviewer 18: ```suggestion
individuals in understanding what a message event is, two shortcuts are defined: `m.text` and
```

`m.html`.

These shortcuts are an easier way to describe the `text/plain` and `text/html` mimetypes, as
shown below:

```json5
{
    "type": "m.message",
    "content": {
        "m.text": "i am a fish",
        "m.html": "i am a <strong>fish</strong>"
    }
}
```

If `m.message` is supplied, `m.html` and `m.text` are ignored. If none of the 3 are supplied
then the event should be ignored as unrenderable. When `m.message` isn't supplied, it is
strongly recommended to at least include `m.text`, however `m.html` can appear without `m.text`
present.

**Note**: We specify `m.html` and not other shortcuts because of the prevelance of HTML in
instant messaging communication today. We could in theory add more shortcuts for `m.markdown`
(of some flavour), `m.rtf`, etc however those are not typically transmitted over the wire
in a reliable or common way.

> **Note**: We specifically support these shortcuts to ensure that funding for Matrix is not

Reviewer 2: it's not really the learning experience we're concerned with in this paragraph, though is another point we'd raise.

interrupted due to perceived complexity. While the shortcuts do add technical complexity for
clients, the simplicity of the example shown above helps demonstrate that it's "easy" to
manually send a message into Matrix. Nested objects for "simple" events demonstrates complexity
which can/has affected funding in the past, though are neccessary for the obviously more

> complicated event types (such as polls or RTF events).

Reviewer 13: This adds a lot of complexity for full clients. I don't agree that putting the burden of demo features in clients is reasonable. You are adding quite a bit of complexity with extensible already, why is this the make or break?

I have seen clients too often depend on events specifically sent by Element, so having one client sending `m.html` and one client `{ "mimetype": "text/html", "body": "i am a <strong>fish</strong>" }` will mean that the message won't render in some clients at some points. This is a significant burden on clients that are smaller than Element, since they need to "implement what Element does" instead of following the specification.

There should be one canonical way to send an event and making the experience worse on Matrix so that it is easier to secure funding is terrible. Especially since it just makes sending messages easier, but receiving them significantly harder.

Reviewer 2: this has been discussed at length already, and the points are covered in the MSC or by other threads here.
